---
title: "RNAseq with edgeR"
author: "Jakob Jung"
date: "November 17, 2020"
output:
  pdf_document: default
  html_document:
    df_print: paged
---
Here I perform the downstream analysis of the RNAseq experiment, in which Salmonella was challenged with PNA targeting the acpP gene and different carrier peptides: KFF, RXR and TAT (3 biological replicates of 4 combinations). RNAseq was performed to determine transcriptome changes upon exposure to PNA. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = TRUE)
```

# Packages
I import the packages needed for all analysis. They can all be installed from Bioconductor or CRAN if not statet otherwise:
```{r, message=FALSE}
library(edgeR)
library(circlize)
library(dplyr)
library(ggplot2)
library('RUVSeq')
library(RColorBrewer)
library(oligo)
library(EDASeq)
library(gplots)
library(ggrepel)
library(svglite)
library(ComplexHeatmap)
library(VennDiagram)
library(eulerr)
library(tidyverse)
library(grid)

```


# Data Acquisition  

I generated a tab file containing gene counts after upstream processing. Upstream processing included folowing steps (starting with fastq-files): 
 
 - BBtools for filtering, trimming and mapping 
 - featureCounts for generating a count matrix


Firstly, I import the gene-wise counts:
```{r}
GenewiseCounts <- read.delim(
  "../data/rna_align/counttable.txt",sep = "\t",
  row.names = 1, header = T, comment.char = "#")

dim(GenewiseCounts)
head(GenewiseCounts[,1:6])
```
\hfill\break

I have to change column names, since they include the whole path:
```{r}
gwc <- GenewiseCounts[,5:length(GenewiseCounts[1,])]
pnapat <- "\\.\\.\\.data\\.rna_align\\..*_(S\\d\\d?)_R1_001\\.bam"
colnames (gwc) <- gsub(pnapat,"\\1", colnames(gwc))
colnames (gwc)
```

I also create a facor variable for groups of the sample data manually (from assigning sample codes to condition):
```{r}
test <- rep(c("Water", "KFF_acpP", "KFF_acpP_scrambled", "KFF", "RXR_acpP", "RXR_acpP_scrambled", 
          "RXR", "TAT_acpP", "TAT_acpP_scrambled", "TAT"), 3)
test <- as.factor(test)
test
```

Now that I have the read count dataframe with sample names, I import them into the edgeR environment:
```{r}
y <- DGEList(gwc[,-1], group = test, genes = gwc[,1,drop=FALSE])
options(digits = 3)
head(y$samples)
```


# Filtering
Now I want to filter out Genes which have very low counts across all libraries. 
I do this by creating a cutoff $$\frac {10} {L} $$
where L is the minimum library size in millions. We delete genes that are below the cutoff in at least 2 libraries:
```{r}
L <- min(y$samples$lib.size) / 1000000
cutoff <- 10/L
keep <- rowSums(cpm(y) > cutoff) >= 2 
table(keep)
```
I retain only the unfiltered genes,and delete 519 genes below the threshold:
```{r}
y <- y[keep, , keep.lib.sizes=FALSE]
```

Find highest expressed genes in raw data controls:
```{r}
controls_raw <- y$counts[,c(1,11,21)]
highest <- names(sort(rowMeans(controls_raw),decreasing = T)[1:1000])
hc_raw <- as.data.frame(controls_raw[highest,])

pnames <- read.delim("../data/link_lt_gn.tab", header = F)
rownames(pnames) <- pnames$V2

prefname <- ifelse(rownames(hc_raw) %in% pnames$V2 ,pnames[rownames(hc_raw),]$V1, "" )
prefname <- ifelse(isUnique(prefname), prefname, "")
hc_raw$genename <- ifelse(prefname != "", prefname, rownames(hc_raw))

write.csv(hc_raw,"../analysis/strong_control.csv")
```


# Design matrix
I create a design matrix for the samples:
```{r}
design <- model.matrix(~0+test)
colnames(design) <- levels(test)
rownames(design) <- colnames(y$counts)
design[1:5,]
```


# Normalization

I check how the standard TMM normalization of edgeR performs. I start with calculating normalization factors:
```{r, error=FALSE}
y <- calcNormFactors(y)
y <- estimateDisp(y, design, robust = T)
```

And now I create PCA and RLE plots:
```{r}
colors <- c(c("#f5a2a2", "#f5a2a2", "#f5a2a2", "#5facc9", "#5facc9", 
              "#5facc9", "#91c497", "#91c497", "#91c497", "white"))
pch <- c(24,21,22,24,21,22,24,21,22,21)
lt <- levels(test)[c(2,3,1,5,6,4,8,9,7,10)]

#t <- gsub("acpP", test)

newtest <- factor(test, levels = lt)
newpch <- c(21,22,24,21,22,24,21,22,24,21)
newcols <- colors[c(2,3,1,5,6,4,8,9,7,10)]


plotPCA(cpm(y), col="black", bg=newcols[newtest],  labels=F, pch=newpch[newtest], cex=3)
legend("bottomright", legend = levels(newtest), pch=newpch, 
       cex=1.2, col="black",  pt.bg = newcols, pt.cex = 2)

plotRLE(cpm(y), outline=FALSE, ylim=c(-1, 1), col=colors[as.factor(newtest)],
        main="RLE")
```

You can see that the TMM was succesful (TMM centers the RLE around 0). Also the variability is rather similar for all samples. This can also be seen in the PCA plot, where the samples separating by condition well, and no batch effects are visible.


I save the PCA as SVG:
```{r}
svg("../analysis/PCA_TMM_new.svg")
#kursiv machen (acpP)
newtest_str <- levels(newtest)
nts <- c(expression( paste("KFF_", italic("acpP"))), expression(paste("KFF_", italic("acpP"), "_scr")), expression("KFF"),
         expression( paste("RXR_", italic("acpP"))), expression(paste("RXR_", italic("acpP"), "_scr")), expression("RXR"),
         expression( paste("Tat_", italic("acpP"))), expression(paste("Tat_", italic("acpP"), "_scr")), expression("Tat"),
         expression("Water control"))

plotPCA(cpm(y), col="black", bg=newcols[newtest],  labels=F, pch=newpch[newtest], cex=3)
legend("bottomright", legend = nts, pch=newpch, 
       cex=1.2, col="black",  pt.bg = newcols, pt.cex = 2)
dev.off()
```


# Differential Expression analysis using TMM
Next, I perform differential expression analysis using TMM-normalized dataset:
```{r}
# make a contrast:
con <- makeContrasts(PNAKFF_vs_ctrl = KFF_acpP - Water,
                     PNAKFFscr_vs_ctrl = KFF_acpP_scrambled - Water,
                     KFF_vs_ctrl = KFF - Water, 
                     PNARXR_vs_ctrl = RXR_acpP - Water,
                     PNARXRscr_vs_ctrl = RXR_acpP_scrambled - Water,
                     RXR_vs_ctrl = RXR - Water,
                     PNATAT_vs_ctrl = TAT_acpP - Water,
                     PNATATscr_vs_ctrl = TAT_acpP_scrambled - Water,
                     TAT_vs_ctrl = TAT - Water,
                     levels = design)

fit <- glmQLFit(y, design, robust = TRUE)

res_KFF <- list(PNAKFF = glmQLFTest(fit, contrast = con[,1]), 
            PNAKFFscr = glmQLFTest(fit, contrast = con[,2]),
            KFF = glmQLFTest(fit, contrast = con[,3]))

res_RXR <- list(PNARXR = glmQLFTest(fit, contrast = con[,4]), 
            PNARXRscr = glmQLFTest(fit, contrast = con[,5]),
            RXR = glmQLFTest(fit, contrast = con[,6]))

res_TAT <- list(PNATAT = glmQLFTest(fit, contrast = con[,7]), 
            PNATATscr = glmQLFTest(fit, contrast = con[,8]),
            TAT = glmQLFTest(fit, contrast = con[,9]))


all_res <- list(KFF = res_KFF,RXR = res_RXR, TAT = res_TAT)
```

do same for CPP vs. scrambled controls:
```{r}
con_cpep <- makeContrasts(KFF_vs_PNAKFFscr = KFF - KFF_acpP_scrambled,
                     RXR_vs_PNARXRscr = RXR - RXR_acpP_scrambled,
                     TAT_vs_PNATATscr = TAT - TAT_acpP_scrambled,
                     levels = design)
cpep_res <- list(KFF_vs_KFFscr = glmQLFTest(fit, contrast = con_cpep[,1]), 
            RXR_vs_RXRscr = glmQLFTest(fit, contrast = con_cpep[,2]),
            TAT_vs_TATscr = glmQLFTest(fit, contrast = con_cpep[,3]))

```






We now create MD, BCV and QLDisp plots to access qualiy of data:
```{r}
plotMD(y, main = "MD-plot")
abline(h=0, col="red", lty=2, lwd=2)
plotBCV(y)
plotQLDisp(fit)
```
The quality looks decent.


Now I create a function which makes nice volcano-plots and run it on all the results (all PNA-samples are compared to water control for DE):
```{r}
do_volcano <- function(restab, pointsize = 2, x_limit = F,y_limit=F, show_sig = F, alpha=0.05, 
                       minlogfc=1, title = "Volcano", sRNAS = F, phopq = F) {
  rownames(restab) <- gsub("^([^A-Z].+)" , "italic('\\1')" , rownames(restab))
  g = ggplot(restab) +
  geom_point(
    data = restab,
    aes(x = logFC, y = -log10(FDR)),
    color = "darkgrey",
    cex = pointsize
  ) + theme_bw()+ # change theme to standard black&wite.
  geom_hline(yintercept = -log10(alpha),
             color = "black", linetype = 3) +
  geom_vline(xintercept = c(-minlogfc,minlogfc),
             color = "black", linetype = 3) +
  theme(axis.title.x = element_text(size=20),
        axis.title.y = element_text(size=20),
        axis.text = element_text(size=15, colour = "black"),
        panel.background = element_rect(colour = "black"),
        axis.line = element_line(colour = "black"),
        panel.grid.minor.x = element_blank(),
        panel.grid.minor.y = element_blank(),
        panel.grid.major.x =  element_blank(),#element_line(colour="lightgrey", size=0.3),
        panel.grid.major.y = element_blank(),#element_line(colour="lightgrey", size=0.3),
        plot.title = element_text(hjust = 0.5, size = 23))+
  ggtitle(title)+
  xlab(expression("Log"[2]*" fold change")) +
  ylab("- Log10 P-value (FDR)")+
  scale_x_continuous(expand = c(0,0),breaks = seq(-6,6,2), limits = c(-x_limit,x_limit)) +
  scale_y_continuous(expand = c(0, 0),breaks = seq(0,16,2), limits = c(0,y_limit)) 

  
  if (sRNAS == F) {
    g <- g + 
      geom_point(
        data = restab[restab$FDR<alpha & restab$logFC < -minlogfc,],
        aes(x = logFC, y = -log10(FDR)),
        color = "blue", 
        cex = pointsize) +
      geom_point(
        data = restab[restab$FDR<alpha & restab$logFC > minlogfc,],
        aes(x = logFC, y = -log10(FDR)),
        color = "red", 
        cex = pointsize) 
  } else{
    show <- restab[sRNAs,][which(restab[sRNAs,]$FDR < alpha),]
    g <- g +
    geom_point(
        data = restab[sRNAs,],
        aes(x = logFC, y = -log10(FDR)),
        color = "darkgreen", 
        cex = pointsize) +
    geom_label_repel(
      data = show ,
      aes(x = logFC, y = -log10(FDR), 
          label = rownames(show)),
      hjust = 0.1,
      vjust = 2,
      size = 4, segment.alpha = 0.5,min.segment.length=0, segment.color = "black") 
  }
  g <- g + 
      geom_point(
        data = restab[restab$genes %in% c("acpP", "fabF"),],
        aes(x = logFC, y = -log10(FDR)),
        bg = "aquamarine3", 
        cex = pointsize+1, pch = 21)
  if (phopq != F) {
    g <- g + geom_point(
        data = restab[restab$genes %in% phopq,],
        aes(x = logFC, y = -log10(FDR)),
        bg = "darkred", 
        cex = pointsize+1, pch=21)
  }
  
  # show the sigficantest genes:
  if(show_sig){
    range01 <- function(x){(x-min(x))/(max(x)-min(x))}
    top_up <- restab[ which(restab$FDR < alpha & restab$logFC > minlogfc),]
    top_down <- restab[ which(restab$FDR < alpha & restab$logFC < -(minlogfc)),]
    
    if (length(rownames(top_up)) > 0 && (length(rownames(top_up)) > 3)){
    logFC.scaled <- range01(top_up$logFC)
    FDR.scaled <- range01(-log(top_up$FDR))
    summ <- (logFC.scaled + FDR.scaled)
    top_up <- top_up[order(-summ),][1:3,]
    }

    if (length(rownames(top_down))>0 && (length(rownames(top_down))> 3)){
      logFC.scaled <- range01(-top_down$logFC)
      FDR.scaled <- range01(-log(top_down$FDR))
      summ <- (logFC.scaled + FDR.scaled)
      top_down <- top_down[order(-summ),][1:3,]
    }

    top_peaks <- rbind(top_up, top_down)
    top_peaks <- na.omit(top_peaks)

    g <- g + geom_label_repel(
    data = top_peaks ,
    aes(x = logFC, y = -log10(FDR), 
        label = rownames(top_peaks)),
    hjust = 0.1,
    #vjust = 2, 
    size = 5, segment.alpha = 0.5, segment.color = "black", min.segment.length=unit(0, "cm"), parse = T) 
  }
  
  
  g
}
```


Now I adjust p-values (FDR), create volcano plots, histograms for the results (and save volcano plots as pdfs): 
```{r}
# I create a variable containing strings of all sRNAs:
sRNAs <- c(rownames(res_KFF$PNAKFF$table)[!grepl("SL1344_", rownames(res_KFF$PNAKFF$table))], "cpxP")

# I get the links between locus tags and gene names:
pnames <- read.delim("../data/link_lt_gn.tab", header = F)
rownames(pnames) <- pnames$V2

# I also import PhoPQ related genes:
ppq_raw <- read.delim("../data/PHOPQ.tsv", header = T)
ppq <- as.character(ppq_raw$PhoPQ[ppq_raw$PhoPQ != ""])
phopqvolc <- c(pnames[pnames$V1 %in% ppq,]$V2, "PinT", "SL1344_1169", "SL1344_1168")
prefname <- ifelse(phopqvolc %in% pnames$V2 ,pnames[phopqvolc,]$V1, "" )
prefname <- ifelse(isUnique(prefname), prefname, "")
phopqvolc <- ifelse(prefname != "", prefname, phopqvolc)
phopqvolc <- c(phopqvolc, "phoP", "phoQ")

for (resname in names(all_res)){
  for (name in names(all_res[[resname]])){
    # adjust p-values FDR
  all_res[[resname]][[name]]$table$FDR <- p.adjust(all_res[[resname]][[name]]$table$PValue, method = "fdr")
  restab <- all_res[[resname]][[name]]$table
  
  #add genenames (not locustags)
  prefname <- ifelse(rownames(restab) %in% pnames$V2 ,pnames[rownames(restab),]$V1, "" )
  prefname <- ifelse(isUnique(prefname), prefname, "")
  rownames(restab) <- ifelse(prefname != "", prefname, rownames(restab))
  
  restab$genes <- rownames(restab)
  
  hist(restab$PValue, breaks=100, main=paste(name," - noPNA"))
  
  # make volcanos:
  pdf(paste("../analysis/volcanoplots/",name, ".pdf"))
  print(do_volcano(restab, title=paste(name," - noPNA"), 
                   x_limit = 7,
                   y_limit = 16,
                   alpha=0.001, pointsize = 3, show_sig = T, phopq = phopqvolc))
  dev.off()
  }
}
```

do volcanos for peptide vs scrambled:
```{r}
for (name in names(cpep_res)){
  # adjust p-values FDR
cpep_res[[name]]$table$FDR <- p.adjust(cpep_res[[name]]$table$PValue, method = "fdr")
restab <- cpep_res[[name]]$table

#add genenames (not locustags)
prefname <- ifelse(rownames(restab) %in% pnames$V2 ,pnames[rownames(restab),]$V1, "" )
prefname <- ifelse(isUnique(prefname), prefname, "")
#rownames(restab) <- ifelse(prefname != "", prefname, rownames(restab))
restab$genename <- ifelse(prefname != "", prefname, rownames(restab))
#restab$gene <- rownames(restab)

hist(restab$PValue, breaks=100, main=paste(name," - noPNA"))

# make volcanos:
pdf(paste("../analysis/volcanoplots/",name, ".pdf"))
print(do_volcano(restab, title=name, 
                 x_limit = 7,
                 y_limit = 16,
                 alpha=0.001, pointsize = 3, show_sig = T, phopq = phopqvolc))
dev.off()

# write csv file:

dataname <- paste("../analysis/", name, ".csv", sep = "")
write.csv(restab[order(restab$FDR),], dataname)
}

```



The volcano-plots that are created look fine, and also the p-value distributions for the samples are uniform with an increment in the lowest p-values showing DE genes.



# Heatmap
I create a heatmap showing the 20 highest ranking DE genesper conditon. I start with the KFF samples:

## KFF
```{r}

for (name in names(res_KFF)) {
  res_KFF[[name]]$table$p_value_FDR <- p.adjust(res_KFF[[name]]$table$PValue, method = "fdr")
  dataname <- paste("../analysis/", name, ".csv", sep = "")
  #write.csv(res_KFF[[name]]$table, dataname)
}

pttk <- res_KFF$PNAKFF$table[order(res_KFF$PNAKFF$table$PValue),]
ptscrk <- res_KFF$PNAKFFscr$table[order(res_KFF$PNAKFFscr$table$PValue),]
ttk <- res_KFF$KFF$table[order(res_KFF$KFF$table$PValue),]

topDEgenes <- c(rownames(pttk[pttk$p_value_FDR<0.001 & 
                                           abs(pttk$logFC)>1,])[1:20],
                rownames(ptscrk[ptscrk$p_value_FDR<0.001 & 
                                           abs(ptscrk$logFC)>1,])[1:20],
                rownames(ttk[ttk$p_value_FDR<0.001 & 
                                           abs(ttk$logFC)>1,])[1:20], 
                "SL1344_1133", "SL1344_1134")
topDEgenes <- unique(topDEgenes[!is.na(topDEgenes)])


logCPM <- cpm(y, prior.count = 2, log = TRUE) 
logCPM <- logCPM[rownames(logCPM)%in%topDEgenes,]
logCPM <- t(scale(t(logCPM))) #centered around 0

prefname <- ifelse(rownames(logCPM) %in% pnames$V2 ,pnames[rownames(logCPM),]$V1, "" )
prefname <- ifelse(isUnique(prefname), prefname, "")
rownames(logCPM) <- ifelse(prefname != "", prefname, rownames(logCPM))

KFF_nrs <- c(1,2,3,4,11,12,13,14,21,22,23,24)
logCPM <- logCPM[,KFF_nrs]

colnames(logCPM) <- factor(c("Untreated R1 ", "on-target PPNA R1 ","scrambled PPNA R1 ","peptide alone R1 ",
                             "Untreated R2 ", "on-target PPNA R2 ","scrambled PPNA R2 ","peptide alone R2 ",
                             "Untreated R3 ", "on-target PPNA R3 ","scrambled PPNA R3 ","peptide alone R3 "))

head(logCPM)
```
Interestingly, the patterns of the samples with scrambled PNA and the test sample (PNA11 and 10 resp.) show similar patterns of the DE genes and cluster together closely.

```{r}
# get log2change, between those and noPNA:
logchange <- data.frame(PNAKFF = res_KFF$PNAKFF$table$logFC,PNAKFFscr =res_KFF$PNAKFFscr$table$logFC,
                              KFF =res_KFF$KFF$table$logFC, row.names = rownames(res_KFF$PNAKFF$table))

pvals <- data.frame(PNAKFF = (res_KFF$PNAKFF$table$p_value_FDR<0.001 & abs(res_KFF$PNAKFF$table$logFC)>1),
                        PNAKFFscr=(res_KFF$PNAKFFscr$table$p_value_FDR<0.001 &
                                     abs(res_KFF$PNAKFFscr$table$logFC)>1),
                        KFF =(res_KFF$KFF$table$p_value_FDR<0.001 & abs(res_KFF$KFF$table$logFC)>1),
                        row.names = rownames(res_KFF$PNAKFF$table))
#select only significant ones:
logchange <- logchange[rownames(logchange)%in%topDEgenes,]

#add genenames
prefname <- ifelse(rownames(logchange) %in% pnames$V2 ,pnames[rownames(logchange),]$V1, "" )
prefname <- ifelse(isUnique(prefname), prefname, "")
rownames(logchange) <- ifelse(prefname != "", prefname, rownames(logchange))

#select only significant ones:
pvals <- pvals[rownames(pvals)%in%topDEgenes,]
#rownames(pvals) <- x$Prefname
colnames(pvals) <- factor(c("on-target PPNA ", "scrambled PPNA ", "peptide alone "))
pvals <-sapply(pvals, function(x) ifelse(x, x <- "*",x<-"") )


prefname <- ifelse(rownames(pvals) %in% pnames$V2 ,pnames[rownames(pvals),]$V1, "" )
prefname <- ifelse(isUnique(prefname), prefname, "")
rownames(pvals) <- ifelse(prefname != "", prefname, rownames(pvals))


colnames(logchange) <- factor(c("on-target PPNA ", "scrambled PPNA ", "peptide alone "))

pvals <- pvals[order(logchange$`on-target PPNA `, decreasing = T),]
logCPM <-logCPM[order(logchange$`on-target PPNA `, decreasing = T),]
logchange <- logchange[order(logchange$`on-target PPNA `, decreasing = T),]

logCPM_T <- t(logCPM[rownames(logchange),])
logchange_T <- t(logchange)
pvals_T <- t(pvals)
rownames(logchange) <- gsub("(.*)", " \\1",rownames(logchange))
```



Now I save the heatmap as pdf:
```{r}

ord2 <- c(rep("on-target PPNA", 3), rep("peptide alone", 3), rep("scrambled PPNA", 3),
         rep("untreated", 3))
lev2 <- c("on-target PPNA", "scrambled PPNA", "peptide alone", "untreated")
logCPM_T <- logCPM_T[sort(rownames(logCPM_T)),]

nr_degenes <- dim(logCPM_T)[2]

c1 =  circlize::colorRamp2(c(-2, 0, 2), c("blue", "white", "red"))
c2 = circlize::colorRamp2(c(-2, 0, 2), c("darkblue", "white", "green"))

ht3 <- Heatmap(logCPM_T, cluster_rows = F, name = "Log CPM",
               show_row_names = F,col=c1,
               show_heatmap_legend = F, cluster_columns = F,
               row_title_side = "right", row_title_rot = 0,
               border = TRUE, 
               column_names_max_height=max_text_width(colnames(logCPM)),
               row_split = factor(ord2, levels = lev2), 
               row_gap = unit(0, "cm"),
               width = unit(nr_degenes, "cm"), height = unit(10, "cm"),
               column_names_rot = 45)
               #rect_gp = gpar(col = "black", lwd = 1))


ht4 <- Heatmap(logchange_T, name = "Log2 FC",
               col = c2,
               cluster_rows = F, cluster_columns = F, show_heatmap_legend = F,
               cell_fun = function(j, i, x, y, width, height, fill) {
                 grid.text(sprintf("%.1s", pvals_T[i, j]), x, y)
               }, 
               border = TRUE, height = unit(3, "cm"),
               row_names_max_width = max_text_width(c(0,0),gp = gpar(fontsize = 0)),
               column_names_rot = 45)
               #rect_gp = gpar(col = "black", lwd = 1))

ht_list = ht3 %v% ht4 
ht_list





lgd1 = Legend(col_fun = c1, title = expression("Log CPM"), labels_gp = gpar(fontsize = 10),
             title_gp = gpar(fontsize = 15),
             at = c(-2, 0, 2), legend_width = unit(4, "cm"), grid_width =  unit(0.8, "cm"),
             labels = c("-2", "  0", "  2"), legend_height = unit(3, "cm"),
             title_position = "leftcenter-rot")
lgd2 = Legend(col_fun = c2, title = expression("Log"[2]*" FC"), labels_gp = gpar(fontsize = 10),
             title_gp = gpar(fontsize = 15),grid_width =  unit(0.8, "cm"),
             at = c(-2, 0, 2), legend_width = unit(4, "cm"),
             labels = c("-2", "  0", "  2"), legend_height = unit(3, "cm"),
             title_position = "leftcenter-rot")

pdf("../analysis/heatmaps/heatmap_KFF_reduced.pdf", width = 20, height = 10)
draw(ht_list, ht_gap = unit(0.5, "cm"))
draw(lgd1, x = unit(5, "cm"), y = unit(14.35, "cm"), just = c("left", "bottom"))
draw(lgd2, x = unit(5, "cm"), y = unit(7.3, "cm"), just = c("left", "bottom"))
dev.off()

```

## RXR heatmap:

I do the same for RXR (redundant):
```{r}

for (name in names(res_RXR)) {
  res_RXR[[name]]$table$p_value_FDR <- p.adjust(res_RXR[[name]]$table$PValue, method = "fdr")
  dataname <- paste("../analysis/", name, ".csv", sep = "")
  #write.csv(res_RXR[[name]]$table, dataname)
}

pttr <- res_RXR$PNARXR$table[order(res_RXR$PNARXR$table$PValue),]
ptscrr <- res_RXR$PNARXRscr$table[order(res_RXR$PNARXRscr$table$PValue),]
ttr <- res_RXR$RXR$table[order(res_RXR$RXR$table$PValue),]

topDEgenes <- c(rownames(pttr[pttr$p_value_FDR<0.001 & 
                                           abs(pttr$logFC)>1,])[1:20],
                rownames(ptscrr[ptscrr$p_value_FDR<0.001 & 
                                           abs(ptscrr$logFC)>1,])[1:20],
                rownames(ttr[ttr$p_value_FDR<0.001 & 
                                           abs(ttr$logFC)>1,])[1:20], 
                "SL1344_1133", "SL1344_1134")
topDEgenes <- unique(topDEgenes[!is.na(topDEgenes)])


logCPM <- cpm(y, prior.count = 2, log = TRUE) 
logCPM <- logCPM[rownames(logCPM)%in%topDEgenes,]
logCPM <- t(scale(t(logCPM))) #centered around 0

#prefname <- ifelse(rownames(logCPM) %in% rownames(GO),GO[rownames(logCPM),]$Preferred_name, "" )
#rownames(logCPM) <- ifelse(prefname != "", prefname, rownames(logCPM))

RXR_nrs <- c(1,5,6,7,11,15,16,17,21,25,26,27)
logCPM <- logCPM[,RXR_nrs]

colnames(logCPM) <- factor(c("Untreated R1 ", "on-target PPNA R1 ","scrambled PPNA R1 ","peptide alone R1 ",
                             "Untreated R2 ", "on-target PPNA R2 ","scrambled PPNA R2 ","peptide alone R2 ",
                             "Untreated R3 ", "on-target PPNA R3 ","scrambled PPNA R3 ","peptide alone R3 "))

head(logCPM)
```

```{r}
library(ComplexHeatmap)
# get log2change, between those and noPNA:
logchange <- data.frame(PNARXR = res_RXR$PNARXR$table$logFC,PNARXRscr =res_RXR$PNARXRscr$table$logFC,
                              RXR =res_RXR$RXR$table$logFC, row.names = rownames(res_RXR$PNARXR$table))

pvals <- data.frame(PNARXR = (res_RXR$PNARXR$table$p_value_FDR<0.001 & abs(res_RXR$PNARXR$table$logFC)>1),
                        PNARXRscr=(res_RXR$PNARXRscr$table$p_value_FDR<0.001 & abs(res_RXR$PNARXRscr$table$logFC)>1),
                        RXR =(res_RXR$RXR$table$p_value_FDR<0.001 & abs(res_RXR$RXR$table$logFC)>1),
                        row.names = rownames(res_RXR$PNARXR$table))
#select only significant ones:
logchange <- logchange[rownames(logchange)%in%topDEgenes,]

colnames(logchange) <- factor(c("RXR-PNA ", "RXR-PNA-scr ", "RXR "))

pnames <- read.delim("../data/link_lt_gn.tab", header = F)
rownames(pnames) <- pnames$V2

prefname <- ifelse(rownames(logchange) %in% pnames$V2 ,pnames[rownames(logchange),]$V1, "" )
prefname <- ifelse(isUnique(prefname), prefname, "")
rownames(logchange) <- ifelse(prefname != "", prefname, rownames(logchange))



#select only significant ones:
pvals <- pvals[rownames(pvals)%in%topDEgenes,]
#rownames(pvals) <- x$Prefname
colnames(pvals) <- factor(c("on-target PPNA ", "scrambled PPNA ", "peptide alone "))
pvals <-sapply(pvals, function(x) ifelse(x, x <- "*",x<-"") )

#prefname <- ifelse(rownames(pvals) %in% rownames(GO),GO[rownames(pvals),]$Preferred_name, "" )
#rownames(pvals) <- ifelse(prefname != "", prefname, rownames(pvals))


colnames(logchange) <- factor(c("on-target PPNA ", "scrambled PPNA ", "peptide alone "))

pvals <- pvals[order(logchange$`on-target PPNA `, decreasing = T),]
logCPM <-logCPM[order(logchange$`on-target PPNA `, decreasing = T),]
logchange <- logchange[order(logchange$`on-target PPNA `, decreasing = T),]

logCPM_T <- t(logCPM)
logchange_T <- t(logchange)
pvals_T <- t(pvals)

rownames(logchange) <- gsub("(.*)", " \\1",rownames(logchange))
```


Now I save the heatmap as pdf:
```{r}

ord2 <- c(rep("on-target PPNA", 3), rep("peptide alone", 3), rep("scrambled PPNA", 3),
         rep("untreated", 3))
lev2 <- c("on-target PPNA", "scrambled PPNA", "peptide alone", "untreated")
logCPM_T <- logCPM_T[sort(rownames(logCPM_T)),]

nr_degenes <- dim(logCPM_T)[2]


ht3 <- Heatmap(logCPM_T, cluster_rows = F, name = "Log CPM",
               show_row_names = F,col=c1,
               show_heatmap_legend = F, cluster_columns = F,
               row_title_side = "right", row_title_rot = 0,
               border = TRUE, 
               column_names_max_height=max_text_width(colnames(logCPM)),
               row_split = factor(ord2, levels = lev2), 
               row_gap = unit(0, "cm"),
               width = unit(nr_degenes, "cm"), height = unit(10, "cm"),
               column_names_rot = 45)
               #rect_gp = gpar(col = "black", lwd = 1))


ht4 <- Heatmap(logchange_T, name = "Log2 FC",
               col = circlize::colorRamp2(c(-2, 0, 2), c("darkblue", "white", "green")),
               cluster_rows = F, cluster_columns = F, show_heatmap_legend = F,
               cell_fun = function(j, i, x, y, width, height, fill) {
                 grid.text(sprintf("%.1s", pvals_T[i, j]), x, y)
               }, 
               border = TRUE, height = unit(3, "cm"),
               row_names_max_width = max_text_width(c(0,0),gp = gpar(fontsize = 0)),
               column_names_rot = 45)
               #rect_gp = gpar(col = "black", lwd = 1))

ht_list = ht3 %v% ht4 
ht_list
pdf("../analysis/heatmaps/heatmap_RXR_reduced.pdf", width = 28, height = 10)
draw(ht_list, ht_gap = unit(0.5, "cm"))
draw(lgd1, x = unit(5, "cm"), y = unit(14.3, "cm"), just = c("left", "bottom"))
draw(lgd2, x = unit(5, "cm"), y = unit(7.3, "cm"), just = c("left", "bottom"))
dev.off()

```


## TAT:
i do the same for TAT
```{r}

for (name in names(res_TAT)) {
  res_TAT[[name]]$table$p_value_FDR <- p.adjust(res_TAT[[name]]$table$PValue, method = "fdr")
  dataname <- paste("../analysis/", name, ".csv", sep = "")
  #write.csv(res_TAT[[name]]$table, dataname)
}


pttt <- res_TAT$PNATAT$table[order(res_TAT$PNATAT$table$PValue),]
ptscrt <- res_TAT$PNATATscr$table[order(res_TAT$PNATATscr$table$PValue),]
ttt <- res_TAT$TAT$table[order(res_TAT$TAT$table$PValue),]


topDEgenes <- c(rownames(pttt[pttt$p_value_FDR<0.001 & abs(pttt$logFC)>1,])[1:20],
                rownames(ptscrt[ptscrt$p_value_FDR<0.001 & 
                                           abs(ptscrt$logFC)>1,])[1:20],
                rownames(ttt[ttt$p_value_FDR<0.001 & 
                                           abs(ttt$logFC)>1,])[1:20], 
                "SL1344_1133", "SL1344_1134")
topDEgenes <- unique(topDEgenes[!is.na(topDEgenes)])



logCPM <- cpm(y, prior.count = 2, log = TRUE) 
logCPM <- logCPM[rownames(logCPM)%in%topDEgenes,]
logCPM <- t(scale(t(logCPM))) #centered around 0
logCPM <- logCPM[topDEgenes,]


TAT_nrs <- c(1,8,9,10,11,18,19,20,21,28,29,30)
logCPM <- logCPM[,TAT_nrs]

colnames(logCPM) <- factor(c("Untreated R1 ", "on-target PPNA R1 ","scrambled PPNA R1 ","peptide alone R1 ",
                             "Untreated R2 ", "on-target PPNA R2 ","scrambled PPNA R2 ","peptide alone R2 ",
                             "Untreated R3 ", "on-target PPNA R3 ","scrambled PPNA R3 ","peptide alone R3 "))

head(logCPM)
```

```{r}
library(ComplexHeatmap)
# get log2change, between those and noPNA:
logchange <- data.frame(PNATAT = res_TAT$PNATAT$table$logFC,PNATATscr =res_TAT$PNATATscr$table$logFC,
                              TAT =res_TAT$TAT$table$logFC, row.names = rownames(res_TAT$PNATAT$table))

pvals <- data.frame(PNATAT = (res_TAT$PNATAT$table$p_value_FDR<0.001 & abs(res_TAT$PNATAT$table$logFC)>1),
                        PNATATscr=(res_TAT$PNATATscr$table$p_value_FDR<0.001 & abs(res_TAT$PNATATscr$table$logFC)>1),
                        TAT =(res_TAT$TAT$table$p_value_FDR<0.001 & abs(res_TAT$TAT$table$logFC)>1),
                        row.names = rownames(res_TAT$PNATAT$table))
#select only significant ones:
logchange <- logchange[rownames(logchange)%in%topDEgenes,]
logchange <- logchange[topDEgenes,]

colnames(logchange) <- factor(c("TAT-PNA ", "TAT-PNA-scr ", "TAT "))


prefname <- ifelse(rownames(logchange) %in% pnames$V2 ,pnames[rownames(logchange),]$V1, "" )
prefname <- ifelse(isUnique(prefname), prefname, "")
rownames(logchange) <- ifelse(prefname != "", prefname, rownames(logchange))




#select only significant ones:
pvals <- pvals[rownames(pvals)%in%topDEgenes,]
pvals <- pvals[topDEgenes,]

#rownames(pvals) <- x$Prefname
colnames(pvals) <- factor(c("on-target PPNA ", "scrambled PPNA ", "peptide alone "))
pvals <-sapply(pvals, function(x) ifelse(x, x <- "*",x<-"") )



colnames(logchange) <- factor(c("on-target PPNA ", "scrambled PPNA ", "peptide alone "))

pvals <- pvals[order(logchange$`on-target PPNA `, decreasing = T),]
logCPM <-logCPM[order(logchange$`on-target PPNA `, decreasing = T),]
logchange <- logchange[order(logchange$`on-target PPNA `, decreasing = T),]

logCPM_T <- t(logCPM)
logchange_T <- t(logchange)
pvals_T <- t(pvals)

rownames(logchange) <- gsub("(.*)", " \\1",rownames(logchange))
```


Now I save the heatmap as pdf:
```{r}

ord2 <- c(rep("on-target PPNA", 3), rep("peptide alone", 3), rep("scrambled PPNA", 3),
         rep("untreated", 3))
lev2 <- c("on-target PPNA", "scrambled PPNA", "peptide alone", "untreated")
logCPM_T <- logCPM_T[sort(rownames(logCPM_T)),]

nr_degenes <- dim(logCPM_T)[2]

ht3 <- Heatmap(logCPM_T, cluster_rows = F, name = "Log CPM",
               show_row_names = F, col=c1,
               show_heatmap_legend = F, cluster_columns = F,
               row_title_side = "right", row_title_rot = 0,
               border = TRUE, 
               column_names_max_height=max_text_width(colnames(logCPM)),
               row_split = factor(ord2, levels = lev2), 
               row_gap = unit(0, "cm"),
               width = unit(nr_degenes, "cm"), height = unit(10, "cm"),
               column_names_rot = 45)
               #rect_gp = gpar(col = "black", lwd = 1))


ht4 <- Heatmap(logchange_T, name = "Log2 FC",
               col = c2,
               cluster_rows = F, cluster_columns = F, show_heatmap_legend = F,
               cell_fun = function(j, i, x, y, width, height, fill) {
                 grid.text(sprintf("%.1s", pvals_T[i, j]), x, y)
               }, 
               border = TRUE, height = unit(3, "cm"),
               row_names_max_width = max_text_width(c(0,0),gp = gpar(fontsize = 0)),
               column_names_rot = 45)
               #rect_gp = gpar(col = "black", lwd = 1))

ht_list = ht3 %v% ht4 
ht_list
pdf("../analysis/heatmaps/heatmap_TAT_reduced.pdf", width = 28, height = 10)
draw(ht_list, ht_gap = unit(0.5, "cm"))
draw(lgd1, x = unit(5, "cm"), y = unit(14.3, "cm"), just = c("left", "bottom"))
draw(lgd2, x = unit(5, "cm"), y = unit(7.3, "cm"), just = c("left", "bottom"))
dev.off()

```

# sRNAs heatmap:
I also create a heatmap for DE sRNAs only to find enriched/reduced sRNAs:
```{r}
topDE <- c(rownames(res_TAT$PNATAT$table[res_TAT$PNATAT$table$p_value_FDR<0.001 & 
                                           abs(res_TAT$PNATAT$table$logFC)>1,]),
                rownames(res_TAT$PNATATscr$table[res_TAT$PNATATscr$table$p_value_FDR<0.001 & 
                                           abs(res_TAT$PNATATscr$table$logFC)>1,]),
                rownames(res_TAT$TAT$table[res_TAT$TAT$table$p_value_FDR<0.001 & 
                                           abs(res_TAT$TAT$table$logFC)>1,]),
                rownames(res_RXR$PNARXR$table[res_RXR$PNARXR$table$p_value_FDR<0.001 & 
                                           abs(res_RXR$PNARXR$table$logFC)>1,]),
                rownames(res_RXR$PNARXRscr$table[res_RXR$PNARXRscr$table$p_value_FDR<0.001 & 
                                           abs(res_RXR$PNARXRscr$table$logFC)>1,]),
                rownames(res_RXR$RXR$table[res_RXR$RXR$table$p_value_FDR<0.001 & 
                                           abs(res_RXR$RXR$table$logFC)>1,]),
                rownames(res_KFF$PNAKFF$table[res_KFF$PNAKFF$table$p_value_FDR<0.001 & 
                                           abs(res_KFF$PNAKFF$table$logFC)>1,]),
                rownames(res_KFF$PNAKFFscr$table[res_KFF$PNAKFFscr$table$p_value_FDR<0.001 & 
                                           abs(res_KFF$PNAKFFscr$table$logFC)>1,]),
                rownames(res_KFF$KFF$table[res_KFF$KFF$table$p_value_FDR<0.001 & 
                                           abs(res_KFF$KFF$table$logFC)>1,]))
topDE_srnas <- unique(topDE[topDE %in% sRNAs])

logchange_TAT <- data.frame(PNATAT = res_TAT$PNATAT$table$logFC,PNATATscr =res_TAT$PNATATscr$table$logFC,
                              TAT =res_TAT$TAT$table$logFC, row.names = rownames(res_TAT$PNATAT$table))

pvals_TAT <- data.frame(PNATAT = (res_TAT$PNATAT$table$p_value_FDR<0.001 & abs(res_TAT$PNATAT$table$logFC)>1),
                        PNATATscr=(res_TAT$PNATATscr$table$p_value_FDR<0.001 & abs(res_TAT$PNATATscr$table$logFC)>1),
                        TAT =(res_TAT$TAT$table$p_value_FDR<0.001 & abs(res_TAT$TAT$table$logFC)>1),
                        row.names = rownames(res_TAT$PNATAT$table))
#select only significant ones:
logchange_TAT <- logchange_TAT[rownames(logchange_TAT)%in%topDE_srnas,]
logchange_TAT <- logchange_TAT[rownames(logchange_TAT),]
colnames(logchange_TAT) <- factor(c("TAT-acpP ", "TAT-acpP-scrambled ", "TAT "))

prefname <- ifelse(rownames(logchange_TAT) %in% pnames$V2 ,pnames[rownames(logchange_TAT),]$V1, "" )
prefname <- ifelse(isUnique(prefname), prefname, "")
rownames(logchange_TAT) <- ifelse(prefname != "", prefname, rownames(logchange_TAT))


#select only significant ones:
pvals_TAT <- pvals_TAT[rownames(pvals_TAT)%in%topDE_srnas,]
pvals_TAT <- pvals_TAT[rownames(pvals_TAT),]
pvals_TAT <-sapply(pvals_TAT, function(x) ifelse(x, x <- "*",x<-"") )
rownames(pvals_TAT) <- rownames(logchange_TAT)




##RXR
logchange_RXR <- data.frame(PNARXR = res_RXR$PNARXR$table$logFC,PNARXRscr =res_RXR$PNARXRscr$table$logFC,
                              RXR =res_RXR$RXR$table$logFC, row.names = rownames(res_RXR$PNARXR$table))

pvals_RXR <- data.frame(PNARXR = (res_RXR$PNARXR$table$p_value_FDR<0.001 & abs(res_RXR$PNARXR$table$logFC)>1),
                        PNARXRscr=(res_RXR$PNARXRscr$table$p_value_FDR<0.001 & abs(res_RXR$PNARXRscr$table$logFC)>1),
                        RXR =(res_RXR$RXR$table$p_value_FDR<0.001 & abs(res_RXR$RXR$table$logFC)>1),
                        row.names = rownames(res_RXR$PNARXR$table))
#select only significant ones:
logchange_RXR <- logchange_RXR[rownames(logchange_RXR)%in%topDE_srnas,]
logchange_RXR <- logchange_RXR[rownames(logchange_RXR),]
colnames(logchange_RXR) <- factor(c("RXR-acpP ", "RXR-acpP-scrambled ", "RXR "))

prefname <- ifelse(rownames(logchange_RXR) %in% pnames$V2 ,pnames[rownames(logchange_RXR),]$V1, "" )
prefname <- ifelse(isUnique(prefname), prefname, "")
rownames(logchange_RXR) <- ifelse(prefname != "", prefname, rownames(logchange_RXR))


#select only significant ones:
pvals_RXR <- pvals_RXR[rownames(pvals_RXR)%in%topDE_srnas,]
pvals_RXR <- pvals_RXR[rownames(pvals_RXR),]
pvals_RXR <-sapply(pvals_RXR, function(x) ifelse(x, x <- "*",x<-"") )
rownames(pvals_RXR) <- rownames(logchange_RXR)




##KFF
logchange_KFF <- data.frame(PNAKFF = res_KFF$PNAKFF$table$logFC,PNAKFFscr =res_KFF$PNAKFFscr$table$logFC,
                              KFF =res_KFF$KFF$table$logFC, row.names = rownames(res_KFF$PNAKFF$table))

pvals_KFF <- data.frame(PNAKFF = (res_KFF$PNAKFF$table$p_value_FDR<0.001 & abs(res_KFF$PNAKFF$table$logFC)>1),
                        PNAKFFscr=(res_KFF$PNAKFFscr$table$p_value_FDR<0.001 & abs(res_KFF$PNAKFFscr$table$logFC)>1),
                        KFF =(res_KFF$KFF$table$p_value_FDR<0.001 & abs(res_KFF$KFF$table$logFC)>1),
                        row.names = rownames(res_KFF$PNAKFF$table))
#select only significant ones:
logchange_KFF <- logchange_KFF[rownames(logchange_KFF)%in%topDE_srnas,]
logchange_KFF <- logchange_KFF[rownames(logchange_KFF),]
colnames(logchange_KFF) <- factor(c("KFF-acpP ", "KFF-acpP-scrambled ", "KFF "))

prefname <- ifelse(rownames(logchange_KFF) %in% pnames$V2 ,pnames[rownames(logchange_KFF),]$V1, "" )
prefname <- ifelse(isUnique(prefname), prefname, "")
rownames(logchange_KFF) <- ifelse(prefname != "", prefname, rownames(logchange_KFF))


#select only significant ones:
pvals_KFF <- pvals_KFF[rownames(pvals_KFF)%in%topDE_srnas,]
pvals_KFF <- pvals_KFF[rownames(pvals_KFF),]

pvals_KFF <-sapply(pvals_KFF, function(x) ifelse(x, x <- "*",x<-"") )
rownames(pvals_KFF) <- rownames(logchange_KFF)


logfcsrnas <- cbind(logchange_KFF, logchange_RXR, logchange_TAT)
pvalssrnas <- cbind(pvals_KFF, pvals_RXR, pvals_TAT)
logfcsrnas <-logfcsrnas[order(logfcsrnas$`KFF-acpP `),] 
pvalssrnas <- pvalssrnas[rownames(logfcsrnas),]


```
Heatmap:
```{r}
col_fun = colorRamp2(c(-2, 0, 2), c("darkblue", "beige", "red"))
ht_vert <- Heatmap(logfcsrnas, cluster_rows = F, cluster_columns = F,
               name = "sRNA", col = col_fun,
               show_heatmap_legend = F, 
               row_title_side = "left", row_title_rot = 0,
               #border = TRUE, 
               cell_fun = function(j, i, x, y, width, height, fill) {
                 grid.text(sprintf("%.1s", pvalssrnas[i, j]), x, y)
               }, 
               column_names_gp = gpar(fontsize = 10),
               row_names_gp = gpar(fontsize = 10),
               column_split = factor(c(rep("KFF",3),rep("RXR",3),rep("TAT",3))), 
               width = unit(9*0.7, "cm"), height = unit(dim(logfcsrnas)[1]/2, "cm"),
               column_names_rot = 45,  border = TRUE)
               #, rect_gp = gpar(col = "black", lwd = 0.01))

lgd = Legend(col_fun = col_fun, title = expression("Log"[2]*" FC"), labels_gp = gpar(fontsize = 8),
             title_gp = gpar(fontsize = 12),
             at = c(-2, 0, 2), legend_width = unit(4, "cm"),
             labels = c("-2", "  0", "  2"), legend_height = unit(3, "cm"),
             title_position = "leftcenter-rot")

pdf("../analysis/heatmaps/heatmap_srnas.pdf")
draw(ht_vert)
draw(lgd, x = unit(2, "cm"), y = unit(8, "cm"), just = c("left", "bottom"))
dev.off()
```




#Venn diagrams:
I create some Venn diagrams (not included in manuscript) to get overview of overlapping DE genes:
```{r}


list_DEgenes <- list(PNA_TAT = rownames(res_TAT$PNATAT$table[res_TAT$PNATAT$table$p_value_FDR<0.001 & 
                                           abs(res_TAT$PNATAT$table$logFC)>1,]),
                PNA_TAT_scr = rownames(res_TAT$PNATATscr$table[res_TAT$PNATATscr$table$p_value_FDR<0.001 & 
                                           abs(res_TAT$PNATATscr$table$logFC)>1,]),
                TAT = rownames(res_TAT$TAT$table[res_TAT$TAT$table$p_value_FDR<0.001 & 
                                           abs(res_TAT$TAT$table$logFC)>1,]),
                PNA_RXR = rownames(res_RXR$PNARXR$table[res_RXR$PNARXR$table$p_value_FDR<0.001 & 
                                           abs(res_RXR$PNARXR$table$logFC)>1,]),
                PNA_RXR_scr = rownames(res_RXR$PNARXRscr$table[res_RXR$PNARXRscr$table$p_value_FDR<0.001 & 
                                           abs(res_RXR$PNARXRscr$table$logFC)>1,]),
                RXR = rownames(res_RXR$RXR$table[res_RXR$RXR$table$p_value_FDR<0.001 & 
                                           abs(res_RXR$RXR$table$logFC)>1,]),
                PNA_KFF = rownames(res_KFF$PNAKFF$table[res_KFF$PNAKFF$table$p_value_FDR<0.001 & 
                                           abs(res_KFF$PNAKFF$table$logFC)>1,]),
                PNA_KFF_scr = rownames(res_KFF$PNAKFFscr$table[res_KFF$PNAKFFscr$table$p_value_FDR<0.001 & 
                                           abs(res_KFF$PNAKFFscr$table$logFC)>1,]),
                KFF = rownames(res_KFF$KFF$table[res_KFF$KFF$table$p_value_FDR<0.001 & 
                                           abs(res_KFF$KFF$table$logFC)>1,]))

svg("../analysis/VennDiagrams/Venn_TMM_TAT",width = 10, height = 10)
plot(euler(list_DEgenes[1:3]) , quantities = T)
dev.off()

svg("../analysis/VennDiagrams/Venn_TMM_RXR",width = 10, height = 10)
plot(euler(list_DEgenes[4:6]) , quantities = T)
dev.off()

svg("../analysis/VennDiagrams/Venn_TMM_KFF",width = 10, height = 10)
plot(euler(list_DEgenes[7:9]) , quantities = T)
dev.off()

svg("../analysis/VennDiagrams/Venn_TMM_peptides",width = 10, height = 10)
plot(euler(list_DEgenes[c(3,6,9)]) , quantities = T)
dev.off()

svg("../analysis/VennDiagrams/Venn_TMM_PPNAs",width = 10, height = 10)
plot(euler(list_DEgenes[c(1,4,7)]) , quantities = T)
dev.off()
```


# KEGG pathway analysis

I perform the KEGG-analysis using the FRY gene set analysis tool from limma:
```{r}
library(KEGGREST)
# get link and list to get kegg info:
link_kegg <- keggLink("pathway", "sey")
list_kegg <- keggList("pathway", "sey")

kegg_pw_ids <- names(list_kegg)

#rename genes, remove ones which arent in our data:
names(link_kegg) <- gsub("sey:(.*)", "\\1", names(link_kegg)) #rename genes as locus tags
link_kegg <- link_kegg[names(link_kegg) %in% rownames(res_KFF$PNAKFF$table)] #remove genes not in data


idx_kegg <- sapply(kegg_pw_ids, function(x){
  x <- unique(names(link_kegg[link_kegg == x])) # choose all genes, except duplucates
})
# add phopq pw to kegg
ppq_raw <- read.delim("../data/PHOPQ.tsv", header = T)

for (c in colnames(ppq_raw)) {
  gs <- ppq_raw[[c]][ppq_raw[[c]]!=""]
  gs_lt <- pnames[pnames$V1 %in% gs,]$V2
  idx_kegg[[c]] <- gs_lt[gs_lt %in% rownames(y$counts)]
}

#ppq <- as.character(ppq_raw$V1)
#phopq <- pnames[pnames$V1 %in% ppq,]$V2

#idx_kegg$PhoPQ <- phopq[phopq %in% rownames(y$counts)] # add PhoPQ genes


#do fry: 
kegg_fry_PNAKFF <- fry(y, idx_kegg, design, contrast=con[,1])
kegg_fry_KFF <- fry(y, idx_kegg, design, contrast=con[,3])
kegg_fry_PNAKFFscr <- fry(y, idx_kegg, design, contrast=con[,2])

kegg_fry_PNARXR <- fry(y, idx_kegg, design, contrast=con[,4])
kegg_fry_RXR <- fry(y, idx_kegg, design, contrast=con[,6])
kegg_fry_PNARXRscr <- fry(y, idx_kegg, design, contrast=con[,5])

kegg_fry_PNATAT <- fry(y, idx_kegg, design, contrast=con[,7])
kegg_fry_TAT <- fry(y, idx_kegg, design, contrast=con[,9])
kegg_fry_PNATATscr <- fry(y, idx_kegg, design, contrast=con[,8])

list_kegg_fry <- list(kegg_fry_PNAKFF=kegg_fry_PNAKFF,kegg_fry_PNAKFFscr=kegg_fry_PNAKFFscr,kegg_fry_KFF=kegg_fry_KFF,
                 kegg_fry_PNARXR=kegg_fry_PNARXR,kegg_fry_PNARXRscr=kegg_fry_PNARXRscr,kegg_fry_RXR=kegg_fry_RXR,
                 kegg_fry_PNATAT=kegg_fry_PNATAT,kegg_fry_PNATATscr=kegg_fry_PNATATscr,kegg_fry_TAT=kegg_fry_TAT)

```

add KEGG terms:
```{r}
for (fryres in names(list_kegg_fry)) {
  list_kegg_fry[[fryres]][["TERM"]] <- ifelse(grepl("path",rownames(list_kegg_fry[[fryres]])),
                                              list_kegg[rownames(list_kegg_fry[[fryres]])],
                                              rownames(list_kegg_fry[[fryres]]))
  list_kegg_fry[[fryres]][["TERM"]] <- gsub("(.*) - Salmonella enterica subsp. enterica serovar Typhimurium SL1344",
                                            "\\1", list_kegg_fry[[fryres]][["TERM"]])
  write.csv(list_kegg_fry[[fryres]], paste("../analysis/analysis_complete/supp_tables/", fryres, ".csv"))
  #list_kegg_fry[[fryres]]["PhoPQ",][["TERM"]] <- "PhoPQ"
}


kegg_frysig <- lapply(list_kegg_fry, function(x) x[x[["FDR"]]<0.05 & x[["NGenes"]]>10,])
kegg_siggos <- c()


for (i in names(kegg_frysig)) {
  print(i)
  print(dim(kegg_frysig[[i]]))
  print(kegg_frysig[[i]][,c(1,2,4,7)])
  kegg_siggos <- c(kegg_siggos, rownames(kegg_frysig[[i]][1:10,]))  # can be modified
}

kegg_siggos <- unique(kegg_siggos[!grepl("NA", kegg_siggos)])


```

Create a heatmap-df  for KEGG:
```{r}
idx_kegg_char <- lapply(idx_kegg, as.character)
# I create a dataframe with mean logFC values for each significant GO-term:
hm_kegg_fry_logfc <- t(as.data.frame(lapply(idx_kegg_char[kegg_siggos], function(x){
  PNAKFF <- median(res_KFF$PNAKFF$table[x,]$logFC)
  PNAKFFscr <- median(res_KFF$PNAKFFscr$table[x,]$logFC)
  KFF <- median(res_KFF$KFF$table[x,]$logFC)
  
  PNARXR <- median(res_RXR$PNARXR$table[x,]$logFC)
  PNARXRscr <- median(res_RXR$PNARXRscr$table[x,]$logFC)
  RXR <- median(res_RXR$RXR$table[x,]$logFC)
  
  PNATAT <- median(res_TAT$PNATAT$table[x,]$logFC)
  PNATATscr <- median(res_TAT$PNATATscr$table[x,]$logFC)
  TAT <- median(res_TAT$TAT$table[x,]$logFC)
  c(PNAKFF, PNAKFFscr, KFF, PNARXR, PNARXRscr, RXR, PNATAT, PNATATscr, TAT)
})))

colnames(hm_kegg_fry_logfc) <- c("KFF-acpP ", "KFF-acpP-scrambled ", "KFF-only ","RXR-acpP ", "RXR-acpP-scrambled ", "RXR-only ",
                            "TAT-acpP ", "TAT-acpP-scrambled ", "TAT-only ")
hm_kegg_fry_logfc <- as.data.frame(hm_kegg_fry_logfc)
rownames(hm_kegg_fry_logfc) <- gsub("\\.", "\\:", rownames(hm_kegg_fry_logfc))
```

make heatmap:
```{r}
hm_kegg_fry_logfc <- hm_kegg_fry_logfc[order(hm_kegg_fry_logfc[,1], decreasing = T),]

pvals <- data.frame(sapply(names(list_kegg_fry), function(x) list_kegg_fry[[x]][rownames(hm_kegg_fry_logfc),"FDR"]), 
                    row.names = rownames(hm_kegg_fry_logfc))

#select only significant ones:
pvals <-sapply(pvals, function(x) ifelse(x<0.05, x <- "*",x<-"") )


keggpws <- list_kegg_fry$kegg_fry_PNAKFF[rownames(hm_kegg_fry_logfc),] [["TERM"]]
rownames(hm_kegg_fry_logfc) <- ifelse(!is.na(keggpws),keggpws, rownames(hm_kegg_fry_logfc) )


ord <- c(rep("KFF", 3), rep("RXR", 3), rep("TAT", 3))
lev <- c("KFF", "RXR", "TAT")
```

plot hm (save as pdf):
```{r}
library(circlize)
col_fun = colorRamp2(c(-1, 0, 1), c("darkblue", "beige", "red"))

nnames <- c("KFF-_acpP_", "KFF-_acpP_-scrambled", "KFF", "RXR-_acpP_", "RXR-_acpP_-scrambled", 
          "RXR", "TAT-_acpP_", "TAT-_acpP_-scrambled", "TAT")

w <- length(hm_kegg_fry_logfc$`KFF-acpP `) # width of plot (nr pws)
h <- 9   # height of plot 

ht_vert <- Heatmap(hm_kegg_fry_logfc, cluster_rows = F, cluster_columns = F,
               name = "GO-analysis", col = col_fun,
               show_heatmap_legend = F, 
               row_title_side = "right", row_title_rot = 0,
               border = TRUE, 
               cell_fun = function(j, i, x, y, width, height, fill) {
                 grid.text(sprintf("%.1s", pvals[i, j]), x, y)
               }, 
               column_names_gp = gpar(fontsize = 10),
               row_names_gp = gpar(fontsize = 10),
               column_split = factor(ord, levels = lev), 
               row_title = NULL,
               row_gap = unit(0.1, "cm"),
               width = unit(5, "cm"), height = unit(20, "cm"),
               column_names_rot = 45)

ht_hor <- Heatmap(t(hm_kegg_fry_logfc), cluster_rows = F, cluster_columns = F,
        name = "GO-analysis", col = col_fun,
        show_heatmap_legend = F, 
        row_title_side = "left", row_title_rot = 0,
        column_names_side = "top",
        border = TRUE, 
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sprintf("%.1s", t(pvals)[i, j]), x, y)
        }, row_labels = gt_render(nnames),
        row_names_gp = gpar(fontsize = 10),
        #column_names_gp = gpar(fontsize = 10),
        row_split = factor(ord, levels = lev), 
        column_title = NULL,
        column_gap = unit(0.1, "cm"),
        width = unit(w*0.8, "cm"), height = unit(h*0.8, "cm"),
        column_names_rot = 40,
        column_names_gp = gpar(col = c("black", rep("darkgreen",2), rep("black", 2), "darkgreen", rep("black", 9),
                                    "darkgreen", "black", "darkgreen", "black", "darkgreen",rep("black", 3)), 
                               fontsize = 10, fontface =  c("plain", rep("bold",2), rep("plain", 2), "bold", rep("plain", 9),
                                    "bold", "plain", "bold", "plain", "bold",rep("plain", 3))
                              )
        )
        #rect_gp = gpar(col = "black", lwd = 0.1))

ht_hor

lgd = Legend(col_fun = col_fun, title = expression("Median logFC"), direction = "horizontal",
             title_gp = gpar(fontsize = 12),
             at = c(-1, 0, 1), legend_width = unit(4, "cm"))

lgd1 = Legend(col_fun = c1, title = expression("Log CPM"), labels_gp = gpar(fontsize = 10),
             title_gp = gpar(fontsize = 15),
             at = c(-2, 0, 2), legend_width = unit(4, "cm"), grid_width =  unit(0.8, "cm"),
             labels = c("-2", "  0", "  2"), legend_height = unit(3, "cm"),
             title_position = "leftcenter-rot")

pdf("../analysis/gene_set_analysis/hm_KEGG_collapsed_new.pdf", width = unit(w*0.45, "cm"))
draw(ht_hor)
draw(lgd, x = unit(w*0.45, "cm"), y = unit(0.8, "cm"), just = c("left", "bottom"))
dev.off()
```


# SPI2 analysis:
Here I perform a SPI2 analysis as described in the manuscript.

First, I read in the dataframes:
```{r}
topDEgenes <- c(rownames(pttt[pttt$p_value_FDR<0.001 & abs(pttt$logFC)>1,])[1:10],
                rownames(ptscrt[ptscrt$p_value_FDR<0.001 & 
                                           abs(ptscrt$logFC)>1,])[1:10],
                rownames(ttt[ttt$p_value_FDR<0.001 & 
                                           abs(ttt$logFC)>1,])[1:10], 
                rownames(pttr[pttr$p_value_FDR<0.001 & abs(pttr$logFC)>1,])[1:10],
                rownames(ptscrr[ptscrr$p_value_FDR<0.001 & 
                                           abs(ptscrr$logFC)>1,])[1:10],
                rownames(ttr[ttr$p_value_FDR<0.001 & 
                                           abs(ttr$logFC)>1,])[1:10],
                rownames(pttk[pttk$p_value_FDR<0.001 & abs(pttk$logFC)>1,])[1:10],
                rownames(ptscrk[ptscrk$p_value_FDR<0.001 & 
                                           abs(ptscrk$logFC)>1,])[1:10],
                rownames(ttk[ttk$p_value_FDR<0.001 & 
                                           abs(ttk$logFC)>1,])[1:10],
                "SL1344_1133", "SL1344_1134")
topDEgenes <- unique(topDEgenes[!is.na(topDEgenes)])
write_delim(data.frame(topDEgenes), "../data/PhoPQ_analysis_salcom/upgenes.txt"," ")

df_spi2 <- read.delim("../data/PhoPQ_analysis_salcom/salcom_query_degenes.txt", header = T, sep="\t")
df_phopq <- read.delim("../data/PhoPQ_analysis_salcom/salcom_query.phopq.txt", header = T, sep="\t")

df_salcom_spi2 <- data.frame(Wildtype = df_spi2$WT.MEP, SPI2 = df_spi2$WT.InSPI2, 
                        PhoPQ_ko = df_spi2$X.Delta.phoP.Q.InSPI2, row.names = df_spi2$SL1344.Locus.ID)

df_salcom_phopq <- data.frame(Wildtype = df_phopq$WT.MEP, SPI2 = df_phopq$WT.InSPI2, 
                        PhoPQ_ko = df_phopq$X.Delta.phoP.Q.InSPI2, row.names = df_phopq$SL1344.Locus.ID)

tpm_salcom_spi2 <- data.frame(sapply(df_salcom_spi2, function(x) as.integer(gsub(",","", x))), 
                              row.names = rownames(df_salcom_spi2))
tpm_salcom_phopq <- data.frame(sapply(df_salcom_phopq, function(x) as.integer(gsub(",","", x))), 
                              row.names = rownames(df_salcom_phopq))

tpm_salcom_spi2 <- tpm_salcom_spi2[order(tpm_salcom_spi2$SPI2),]
tpm_salcom_phopq <- tpm_salcom_phopq[order(tpm_salcom_phopq$SPI2),]
tpm_salcom_spi2 <- tpm_salcom_spi2[!(rownames(tpm_salcom_spi2)=="SL1344_1325"),]

tpm_salcom_spi2 <- tpm_salcom_spi2[!(rownames(tpm_salcom_spi2) %in% rownames(tpm_salcom_phopq)),]

tpm_salcom_spi2$condition <- "DE genes"
tpm_salcom_phopq$condition <- "PhoPQ"



rownames(tpm_salcom_phopq) <- gsub(".*MgrR", "MgrR", rownames(tpm_salcom_phopq))
tpm_salcom <- data.frame(rbind(tpm_salcom_phopq, tpm_salcom_spi2), row.names = c(rownames(tpm_salcom_phopq),
                                                                                 rownames(tpm_salcom_spi2)))

str(tpm_salcom)
```
Get all genes which are included in both datasets:
```{r}
all_spi2_genes <- rownames(tpm_salcom)[rownames(tpm_salcom) %in% rownames(y)]
str(all_spi2_genes)
tpm_salcom <- tpm_salcom[all_spi2_genes,]

prefname <- ifelse(all_spi2_genes %in% pnames$V2 ,pnames[all_spi2_genes,]$V1, "" )
prefname <- ifelse(isUnique(prefname), prefname, "")
prefname_spi2 <- ifelse(prefname != "", prefname, all_spi2_genes)

rownames(tpm_salcom) <- prefname_spi2
```

Now we have to make heatmaps with log foldchanges for all samples:
```{r}
# make list of all results, get logchange table:
reslist <- list(KFF = res_KFF,RXR = res_RXR, TAT = res_TAT)
logchanges_spi2 <- data.frame(lapply(reslist, function(l) {
  logfc <- data.frame( sapply(names(l), function(r) {
    l[[r]]$table[all_spi2_genes,1]
  }), row.names = all_spi2_genes )
})) 

colnames(logchanges_spi2) <- c("KFF-acpP ", "KFF-acpP-scrambled ", "KFF-only ","RXR-acpP ",
                          "RXR-acpP-scrambled ", "RXR-only ",
                          "Tat-acpP ", "Tat-acpP-scrambled ", "Tat-only")


# get mean cpm values of all conditions:
countscpm <- cpm(y)[all_spi2_genes,]

spi2_cpm <- sapply(levels(test), function(t) {
  rowMeans(countscpm[,t == test])
})
spi2_cpm <- spi2_cpm[,c(2,3,1,5,6,4,8,9,7,10)]


pvalues_spi2 <- data.frame(lapply(reslist, function(l) {
  logfc <- data.frame( sapply(names(l), function(r) {
    l[[r]]$table[all_spi2_genes,]$p_value_FDR<0.001 & abs(l[[r]]$table[all_spi2_genes,]$logFC)>1
  }), row.names = all_spi2_genes )
})) 

pvalues_spi2 <-sapply(pvalues_spi2, function(x) ifelse(x , x <- "*",x<-"")) 

```




Plot Heatmap:
```{r}
h <- dim(logchanges_spi2)[1] # width of plot (nr pws)

col_fun = colorRamp2(c(-2, 0, 2), c("blue", "beige", "red"))
ht_vert <- Heatmap(logchanges_spi2, cluster_rows = F, cluster_columns = F,
               name = "SPI2 analysis", col = col_fun,
               show_heatmap_legend = F, 
               row_title_side = "left", row_title_rot = 0,
               border = TRUE, 
               cell_fun = function(j, i, x, y, width, height, fill) {
                 grid.text(sprintf("%.1s", pvalues_spi2[i, j]), x, y)
               }, 
               column_names_gp = gpar(fontsize = 11),
               row_names_gp = gpar(fontsize = 10),
               column_split = factor(ord, levels = lev), 
               row_split = factor(tpm_salcom$condition),
               row_gap = unit(0.2, "cm"),
               width = unit(9*0.8, "cm"), height = unit(h/2, "cm"),
               column_names_rot = 45)


col_col_fun = colorRamp2(c(0, 2,3), c("lightgrey", "yellow", "red"))
ht_colgan <- Heatmap(log10(tpm_salcom[,1:3]), cluster_rows = F, cluster_columns = F,
               name = "Colgan et al., 2016", col = col_col_fun,
               show_heatmap_legend = F, 
               row_title_side = "right", row_title_rot = 0,
               border = TRUE, 
               row_split = factor(tpm_salcom$condition),
               column_names_gp = gpar(fontsize = 11),
               row_names_gp = gpar(fontsize = 10),
               row_gap = unit(0.2, "cm"),
               width = unit(3*0.8, "cm"), height = unit(h/2, "cm"),
               column_names_rot = 45)

order = factor(c(rep("KFF", 3), rep("RXR", 3), rep("TAT", 3), "Control"))



ht_list =  ht_vert + ht_colgan 
ht_list

lgd = Legend(col_fun = col_fun, title = expression("Log"[2]*" FC"),
             at = c(-2, 0, 2), legend_width = unit(4, "cm"),
             title_gp  = gpar(fontsize(14)),
             labels = c("-2", "  0", "  2"), legend_height = unit(3, "cm"),
             title_position = "leftcenter-rot")
lgd_col = Legend(col_fun = col_col_fun, title = expression("TPM"),
             at = c(0, 1, 2, 3), labels = c(0,10,100,1000), legend_width = unit(3, "cm"),
             title_gp  = gpar(fontsize(14)),
             legend_height = unit(3, "cm"),
             title_position = "leftcenter-rot")


pdf("../analysis/PhoPQ_Salcom/hm_salcom_phopq_spi2.pdf", width = 10, height = 15)
draw(ht_list, ht_gap = unit(0.5, "cm"))
draw(lgd, x = unit(4, "cm"), y = unit(18, "cm"), just = c("left", "bottom"))
draw(lgd_col, x = unit(21, "cm"), y = unit(18, "cm"), just = c("left", "bottom"))
dev.off()
```


save csv of all raw counts for supplementary material:
```{r}
counts <- y$counts 
test <- c("Water1", "KFF_acpP1", "KFF_acpP_scrambled1", "KFF1", "RXR_acpP1", "RXR_acpP_scrambled1", 
          "RXR1", "TAT_acpP1", "TAT_acpP_scrambled1", "TAT1",
          "Water2", "KFF_acpP2", "KFF_acpP_scrambled2", "KFF2", "RXR_acpP2", "RXR_acpP_scrambled2", 
          "RXR2", "TAT_acpP2", "TAT_acpP_scrambled2", "TAT2",
          "Water3", "KFF_acpP3", "KFF_acpP_scrambled3", "KFF3", "RXR_acpP3", "RXR_acpP_scrambled3", 
          "RXR3", "TAT_acpP3", "TAT_acpP_scrambled3", "TAT3")
colnames(counts) <- test

prefname <- ifelse(rownames(counts) %in% pnames$V2 ,pnames[rownames(counts),]$V1, "" )
prefname <- ifelse(isUnique(prefname), prefname, "")
gene_name <- ifelse(prefname != "", prefname, rownames(counts))

locus_tag <- rownames(counts)

counts_raw <- cbind(locus_tag, gene_name, counts)

write.csv(counts_raw, "../analysis/analysis_complete/supp_tables/raw_counts.csv")

for (i in all_res) {
  for (l in names(i)){
    tab <- i[[l]]$table[order(i[[l]]$table$FDR),]
    prefname <- ifelse(rownames(tab) %in% pnames$V2 ,pnames[rownames(tab),]$V1, "" )
    prefname <- ifelse(isUnique(prefname), prefname, "")
    genename <- ifelse(prefname != "", prefname, rownames(tab))
    tabs <- cbind(genename,tab) 
    write.csv(tabs, paste("../analysis/analysis_complete/supp_tables/",l,"_vs_water",".csv", sep = ""))
  }
}  


for (l in names(list_kegg_fry)){
  tabs <- list_kegg_fry[[l]]
  write.csv(tabs, paste("../analysis/analysis_complete/supp_tables/",l,"_vs_water",".csv", sep = ""))
}


```


Packages used:
```{r}
sessionInfo()
```

