---
title: "RNAseq with edgeR"
author: "Jakob Jung"
date: "November 17, 2020"
output:
  pdf_document: default
  html_document:
    df_print: paged
---
Here I perform the downstream analysis of the RNAseq experiment, in which Salmonella was challenged with PNA targeting the acpP gene and the carrier peptide KFF (3 biological replicates of 4 combinations). Then RNAseq was performed to determine which genes were silenced and DE between conditions. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = TRUE)
```

# Packages
Here the necessary packages are imported. They can be installed from Bioconductor or CRAN if not statet otherwise:
```{r, message=FALSE}
library(edgeR)
library(circlize)
library(VennDiagram)
library(dplyr)
library(ggplot2)
library('RUVSeq')
library(RColorBrewer)
library(oligo)
library(EDASeq)
library(gplots)
library(ggrepel)
library(svglite)
```



# Data Acquisition  

A tab file containing gene counts was generated after upstream processing. Upstream processing included folowing steps (starting with fastq-files): 
 
 - BBtools pipeliine for filtering, trimming and mapping 
 - HTSeq-count for read-counting 


Let's import the gene-wise counts:
```{r}
GenewiseCounts <- read.delim(
  "../data/rna_align/counttable.txt",sep = "\t",
  row.names = 1, header = T, comment.char = "#")

dim(GenewiseCounts)
head(GenewiseCounts[,1:6])
```
\hfill\break

We have to change colnames, since they include the whole path:
```{r}
gwc <- GenewiseCounts[,5:length(GenewiseCounts[1,])]
pnapat <- "\\.\\.\\.data\\.rna_align\\..*_(S\\d\\d?)_R1_001\\.bam"
colnames (gwc) <- gsub(pnapat,"\\1", colnames(gwc))
colnames (gwc)
```

We also create a variable for groups of the sample data manually (from the colnames):
```{r}
#test <- rep(c("Water control", "KFF-acpP", "KFF-acpP-scr", "KFF", "RXR-acpP", "RXR-acpP-scr", 
#          "RXR", "TAT-acpP", "TAT-acpP-scr", "TAT"), 3)
test <- rep(c("Water", "KFF_acpP", "KFF_acpP_scrambled", "KFF", "RXR_acpP", "RXR_acpP_scrambled", 
          "RXR", "TAT_acpP", "TAT_acpP_scrambled", "TAT"), 3)
test <- as.factor(test)
test
```
The meaning of sample names is:

- #replicate_#sample

Now that we have these two dataframes, we can import them into the edgeR environment:
```{r}
y <- DGEList(gwc[,-1], group = test, genes = gwc[,1,drop=FALSE])
options(digits = 3)
head(y$samples)
```


# Filtering
Now we want to filter out Genes which have very low counts across all libraries. 
We do this by creating a cutoff $$\frac {10} {L} $$
where L is the minimum library size in millions. We delete genes that are below the cutoff in at least 2 libraries:
```{r}
L <- min(y$samples$lib.size) / 1000000
cutoff <- 10/L
keep <- rowSums(cpm(y) > cutoff) >= 2 
table(keep)
```
We retain only the unfiltered genes:
```{r}
y <- y[keep, , keep.lib.sizes=FALSE]
```





# Design matrix
We create a design matrix for the samples:
```{r}
design <- model.matrix(~0+test)
colnames(design) <- levels(test)
rownames(design) <- colnames(y$counts)
design[1:5,]
```


# Normalization
Here, we check different Normalization techniques and access their effectivity.

## TMM Normalization
Let's check how the usually recommended TMM normalization of edgeR performs. for this we use RLE and PCA plots:
```{r}
y <- calcNormFactors(y)
y <- estimateDisp(y, design, robust = T)
colors <- c(brewer.pal(10, "Set1"),"blue")
plotRLE(cpm(y), outline=FALSE, ylim=c(-1, 1), col=colors[as.factor(test)])
plotPCA(cpm(y), col=colors[as.factor(test)], cex=0.7)
```
We can see that the TMM was succesful (TMM centers the RLE around 0). Also the variability is approximately similar for all samples. This can also be seen in the PCA plot, where the samples separating quite well. Anyway, we'll test whether we can get some improvement using the RUV normalization method.


we save the PCA and heatmap:
```{r}
colors <- c(c("#f5a2a2", "#ffe8e8", "#f7c6c6", "#235a71", "#5facc9", 
              "#578ca1", "#71a876", "#dcf2dd", "#91c497", "#d3ebf5"))

pch <- c(24,21,22,24,21,22,24,21,22,21)
#save as pdfs:
#pdf("../analysis/TAT_TMM_PCA_RLE_hm")
plotPCA(cpm(y), col=colors[test], cex=3,labels=F, pch=pch[test], main="PCA")
legend("bottomright", legend = levels(test), pch=pch, col=colors, ncol=2, cex=1)

plotRLE(cpm(y), outline=FALSE, ylim=c(-1, 1), col=colors[as.factor(test)],
        main="RLE")
#dev.off()
lt <- levels(test)[c(2,3,1,5,6,4,8,9,7,10)]
newtest <- factor(test, levels = lt)
newpch <- c(21,22,24,21,22,24,21,22,24,21)
newcols <- colors[c(2,3,1,5,6,4,8,9,7,10)]

svg("../analysis/PCA_TMM.svg")
plotPCA(cpm(y), col="black", bg=newcols[newtest],  labels=F, pch=newpch[newtest], cex=3)
legend("bottomright", legend = levels(newtest), pch=newpch, 
       cex=1.2, col="black",  pt.bg = newcols, pt.cex = 2)
dev.off()


plotRLE(cpm(y), col=colors[test], cex=9,labels=F,  outline=FALSE, ylim=c(-2, 2), xaxt="none")
legend("topleft", legend = levels(test),  col=colors, pch=rep(15,3))

```

Create PCAs for each CPP experiment:
```{r}
KFF_nrs <- c(1,2,3,4,11,12,13,14,21,22,23,24)
RXR_nrs <- c(1,5,6,7,11,15,16,17,21,25,26,27)
TAT_nrs <- c(1,8,9,10,11,18,19,20,21,28,29,30)

test_KFF <- droplevels(test[KFF_nrs])
test_RXR <- droplevels(test[RXR_nrs])
test_TAT <- droplevels(test[TAT_nrs]) 

svg("../analysis/PCA_KFF.svg")
plotPCA(cpm(y)[,KFF_nrs], col=colors[test_KFF], labels=F, pch=pch[test_KFF], cex=2)
legend("bottomright", legend = levels(test_KFF), pch=pch, cex=1, col=colors)
dev.off()

svg("../analysis/PCA_RXR.svg")
plotPCA(cpm(y)[,RXR_nrs], col=colors[test_RXR], labels=F, pch=pch[test_RXR], cex=2)
legend("bottomright", legend = levels(test_RXR), pch=pch, cex=1, col=colors)
dev.off()

svg("../analysis/PCA_TAT.svg")
plotPCA(cpm(y)[,TAT_nrs], col=colors[test_TAT], labels=F, pch=pch[test_TAT], cex=2)
legend("bottomright", legend = levels(test_TAT), pch=pch, cex=1, col=colors)
dev.off()
```



# Differential Expression analysis using TMM
Next, we do differential expression analysis using the TMM-normalized dataset:

```{r}
# make a contrast:
con <- makeContrasts(PNAKFF_vs_ctrl = KFF_acpP - Water,
                     PNAKFFscr_vs_ctrl = KFF_acpP_scrambled - Water,
                     KFF_vs_ctrl = KFF - Water, 
                     PNARXR_vs_ctrl = RXR_acpP - Water,
                     PNARXRscr_vs_ctrl = RXR_acpP_scrambled - Water,
                     RXR_vs_ctrl = RXR - Water,
                     PNATAT_vs_ctrl = TAT_acpP - Water,
                     PNATATscr_vs_ctrl = TAT_acpP_scrambled - Water,
                     TAT_vs_ctrl = TAT - Water,
                     levels = design)

fit <- glmQLFit(y, design, robust = TRUE)

res_KFF <- list(PNAKFF = glmQLFTest(fit, contrast = con[,1]), 
            PNAKFFscr = glmQLFTest(fit, contrast = con[,2]),
            KFF = glmQLFTest(fit, contrast = con[,3]))

res_RXR <- list(PNARXR = glmQLFTest(fit, contrast = con[,4]), 
            PNARXRscr = glmQLFTest(fit, contrast = con[,5]),
            RXR = glmQLFTest(fit, contrast = con[,6]))

res_TAT <- list(PNATAT = glmQLFTest(fit, contrast = con[,7]), 
            PNATATscr = glmQLFTest(fit, contrast = con[,8]),
            TAT = glmQLFTest(fit, contrast = con[,9]))


all_res <- list(KFF = res_KFF,RXR = res_RXR, TAT = res_TAT)
```

We now create MD, BCV and QLDisp plots to access qualiy of data:
```{r}
plotMD(y, main = "MD-plot")
abline(h=0, col="red", lty=2, lwd=2)
plotBCV(y)
plotQLDisp(fit)
```
The quality looks decent.


Now we create a function which makes nice volcano-plots and run it on all the results (all PNA-samples are compared to noPNA for DE):
```{r}


do_volcano <- function(restab, pointsize = 2, x_limit = F,y_limit=F, show_sig = F, alpha=0.05, 
                       minlogfc=1, title = "Volcano", sRNAS = F) {
  g = ggplot(restab) +
  geom_point(
    data = restab,
    aes(x = logFC, y = -log10(FDR)),
    color = "darkgrey",
    cex = pointsize
  ) + theme_bw()+ # chane theme to standard black&wite.
  geom_hline(yintercept = -log10(alpha),
             color = "black", linetype = 3) +
  geom_vline(xintercept = c(-minlogfc,minlogfc),
             color = "black", linetype = 3) +
  theme(axis.title.x = element_text(size=20),
        axis.title.y = element_text(size=20),
        axis.text.x = element_text(size=15),
        axis.text.y = element_text(size=15),
        panel.grid.minor.x = element_blank(),
        panel.grid.minor.y = element_blank(),
        panel.grid.major.x =  element_blank(),#element_line(colour="lightgrey", size=0.3),
        panel.grid.major.y = element_blank(),#element_line(colour="lightgrey", size=0.3),
        plot.title = element_text(hjust = 0.5, size = 23))+
  ggtitle(title)+
  xlab(expression("Log"[2]*" fold change")) +
  ylab("- Log10 P-value (FDR adjusted)")+
  scale_x_continuous(expand = c(0,0),breaks = seq(-6,6,2), limits = c(-x_limit,x_limit)) +
  scale_y_continuous(expand = c(0, 0),breaks = seq(0,16,2), limits = c(0,y_limit)) 

  
  if (sRNAS == F) {
    g <- g + 
      geom_point(
        data = restab[restab$FDR<alpha & restab$logFC < -minlogfc,],
        aes(x = logFC, y = -log10(FDR)),
        color = "blue", 
        cex = pointsize) +
      geom_point(
        data = restab[restab$FDR<alpha & restab$logFC > minlogfc,],
        aes(x = logFC, y = -log10(FDR)),
        color = "red", 
        cex = pointsize) 
  } else{
    show <- restab[sRNAs,][which(restab[sRNAs,]$FDR < alpha),]
    g <- g +
    geom_point(
        data = restab[sRNAs,],
        aes(x = logFC, y = -log10(FDR)),
        color = "darkgreen", 
        cex = pointsize) +
    geom_label_repel(
      data = show ,
      aes(x = logFC, y = -log10(FDR), 
          label = rownames(show)),
      hjust = 0.1,
      vjust = 2,
      size = 4, segment.alpha = 0.5,min.segment.length=0, segment.color = "black") 
  }
  # show the sigficantest genes:
  if(show_sig){
    range01 <- function(x){(x-min(x))/(max(x)-min(x))}
    top_up <- restab[ which(restab$FDR < alpha & restab$logFC > minlogfc),]
    top_down <- restab[ which(restab$FDR < alpha & restab$logFC < -(minlogfc)),]
    
    if (length(rownames(top_up)) > 0 && (length(rownames(top_up)) > 10)){
    logFC.scaled <- range01(top_up$logFC)
    FDR.scaled <- range01(-log(top_up$FDR))
    summ <- (logFC.scaled + FDR.scaled)
    top_up <- top_up[order(-summ),][1:10,]
    }

    if (length(rownames(top_down))>0 && (length(rownames(top_down))> 10)){
      logFC.scaled <- range01(-top_down$logFC)
      FDR.scaled <- range01(-log(top_down$FDR))
      summ <- (logFC.scaled + FDR.scaled)
      top_down <- top_down[order(-summ),][1:10,]
    }

    top_peaks <- rbind(top_up, top_down)
    top_peaks <- na.omit(top_peaks)
    
    g <- g + geom_label_repel(
    data = top_peaks ,
    aes(x = logFC, y = -log10(FDR), 
        label = rownames(top_peaks)),
    hjust = 0.1,
    vjust = 2, 
    size = 4, segment.alpha = 0.5, segment.color = "black", min.segment.length=unit(0, "cm")) 
  }
  
  
  g
}
```

Now we adjust p-values, create volcano plots, histograms for the results (and save as pdfs): 
```{r}
sRNAs <- c(rownames(res_KFF$PNAKFF$table)[!grepl("SL1344_", rownames(res_KFF$PNAKFF$table))], "cpxP")
pnames <- read.delim("../data/link_lt_gn.tab", header = F)
rownames(pnames) <- pnames$V2

for (resname in names(all_res)){
  for (name in names(all_res[[resname]])){
    # adjust p-values FDR
  all_res[[resname]][[name]]$table$FDR <- p.adjust(all_res[[resname]][[name]]$table$PValue, method = "fdr")
  restab <- all_res[[resname]][[name]]$table
  
  #add genenames
  prefname <- ifelse(rownames(restab) %in% pnames$V2 ,pnames[rownames(restab),]$V1, "" )
  prefname <- ifelse(isUnique(prefname), prefname, "")
  rownames(restab) <- ifelse(prefname != "", prefname, rownames(restab))
  
  restab$genes <- rownames(restab)
  #make_volcanoPlot(restab, paste(name,"_noPNA", sep = ""))
  
    #make volcano plot
  svg(paste("../analysis/volcanoplots/",name, "_TMM.svg"))
  print(do_volcano(restab, title=paste(name," - noPNA"), 
                   x_limit = 7,
                   y_limit = 16,
                   alpha=0.001, pointsize = 3, show_sig = T))
  dev.off()
  #hist(restab$PValue, breaks=100, main=paste(name," - noPNA"))
  
  # make sRNA volcanos:
  svg(paste("../analysis/volcanoplots/sRNAs/",name, "_sRNAs.svg"))
  print(do_volcano(restab, title=paste(name," - noPNA"), 
                   x_limit = 7,
                   y_limit = 16,
                   alpha=0.001, pointsize = 3, show_sig = F,
                   sRNAS = sRNAs))
  dev.off()
  }
}
```





The volcano-plots that are created look fine, and also the p-value distribution for the samples is either uniform with an increment in the lowest p-values showing DE genes.



# Heatmap
lets create a heatmap clustering all highest DE genes of all contrasts.
```{r}

for (name in names(res_KFF)) {
  res_KFF[[name]]$table$p_value_FDR <- p.adjust(res_KFF[[name]]$table$PValue, method = "fdr")
  dataname <- paste("../analysis/", name, ".csv", sep = "")
  #write.csv(res_KFF[[name]]$table, dataname)
}

topDEgenes <- c(rownames(res_KFF$PNAKFF$table[res_KFF$PNAKFF$table$FDR<0.001 & 
                                           abs(res_KFF$PNAKFF$table$logFC)>1,])[1:20],
                rownames(res_KFF$PNAKFFscr$table[res_KFF$PNAKFFscr$table$FDR<0.001 & 
                                           abs(res_KFF$PNAKFFscr$table$logFC)>1,])[1:20],
                rownames(res_KFF$KFF$table[res_KFF$KFF$table$FDR<0.001 & 
                                           abs(res_KFF$KFF$table$logFC)>1,])[1:20], 
                "SL1344_1133", "SL1344_1134")
topDEgenes <- unique(topDEgenes[!is.na(topDEgenes)])


logCPM <- cpm(y, prior.count = 2, log = TRUE) 
logCPM <- logCPM[rownames(logCPM)%in%topDEgenes,]
logCPM <- t(scale(t(logCPM))) #centered around 0

#prefname <- ifelse(rownames(logCPM) %in% rownames(GO),GO[rownames(logCPM),]$Preferred_name, "" )
#rownames(logCPM) <- ifelse(prefname != "", prefname, rownames(logCPM))

prefname <- ifelse(rownames(logCPM) %in% pnames$V2 ,pnames[rownames(logCPM),]$V1, "" )
prefname <- ifelse(isUnique(prefname), prefname, "")
rownames(logCPM) <- ifelse(prefname != "", prefname, rownames(logCPM))

logCPM <- logCPM[,KFF_nrs]

colnames(logCPM) <- factor(c("Untreated R1 ", "on-target PPNA R1 ","scrambled PPNA R1 ","peptide alone R1 ",
                             "Untreated R2 ", "on-target PPNA R2 ","scrambled PPNA R2 ","peptide alone R2 ",
                             "Untreated R3 ", "on-target PPNA R3 ","scrambled PPNA R3 ","peptide alone R3 "))

head(logCPM)
```
Interestingly, the patterns of the samples with scrambled PNA and the test sample (PNA11 and 10 resp.) show similar patterns of the DE genes and cluster together closely.
```{r}
library(ComplexHeatmap)
# get log2change, between those and noPNA:
logchange <- data.frame(PNAKFF = res_KFF$PNAKFF$table$logFC,PNAKFFscr =res_KFF$PNAKFFscr$table$logFC,
                              KFF =res_KFF$KFF$table$logFC, row.names = rownames(res_KFF$PNAKFF$table))

pvals <- data.frame(PNAKFF = (res_KFF$PNAKFF$table$p_value_FDR<0.001 & abs(res_KFF$PNAKFF$table$logFC)>1),
                        PNAKFFscr=(res_KFF$PNAKFFscr$table$p_value_FDR<0.001 &
                                     abs(res_KFF$PNAKFFscr$table$logFC)>1),
                        KFF =(res_KFF$KFF$table$p_value_FDR<0.001 & abs(res_KFF$KFF$table$logFC)>1),
                        row.names = rownames(res_KFF$PNAKFF$table))
#select only significant ones:
logchange <- logchange[rownames(logchange)%in%topDEgenes,]

#add genenames
prefname <- ifelse(rownames(logchange) %in% pnames$V2 ,pnames[rownames(logchange),]$V1, "" )
prefname <- ifelse(isUnique(prefname), prefname, "")
rownames(logchange) <- ifelse(prefname != "", prefname, rownames(logchange))

#select only significant ones:
pvals <- pvals[rownames(pvals)%in%topDEgenes,]
#rownames(pvals) <- x$Prefname
colnames(pvals) <- factor(c("on-target PPNA ", "scrambled PPNA ", "peptide alone "))
pvals <-sapply(pvals, function(x) ifelse(x, x <- "*",x<-"") )

#prefname <- ifelse(rownames(pvals) %in% rownames(GO),GO[rownames(pvals),]$Preferred_name, "" )
#rownames(pvals) <- ifelse(prefname != "", prefname, rownames(pvals))
prefname <- ifelse(rownames(pvals) %in% pnames$V2 ,pnames[rownames(pvals),]$V1, "" )
prefname <- ifelse(isUnique(prefname), prefname, "")
rownames(pvals) <- ifelse(prefname != "", prefname, rownames(pvals))


colnames(logchange) <- factor(c("on-target PPNA ", "scrambled PPNA ", "peptide alone "))

pvals <- pvals[order(logchange$`on-target PPNA `, decreasing = T),]
logCPM <-logCPM[order(logchange$`on-target PPNA `, decreasing = T),]
logchange <- logchange[order(logchange$`on-target PPNA `, decreasing = T),]

logCPM_T <- t(logCPM)
logchange_T <- t(logchange)
pvals_T <- t(pvals)
rownames(logchange) <- gsub("(.*)", " \\1",rownames(logchange))
```



Now we save the heatmap as svg file:
```{r}

ord2 <- c(rep("on-target PPNA", 3), rep("peptide alone", 3), rep("scrambled PPNA", 3),
         rep("untreated", 3))
lev2 <- c("on-target PPNA", "scrambled PPNA", "peptide alone", "untreated")
logCPM_T <- logCPM_T[sort(rownames(logCPM_T)),]

nr_degenes <- dim(logCPM_T)[2]

ht3 <- Heatmap(logCPM_T, cluster_rows = F, name = "Log CPM",
               show_row_names = F,
               show_heatmap_legend = T, cluster_columns = F,
               row_title_side = "right", row_title_rot = 0,
               border = TRUE, 
               column_names_max_height=max_text_width(colnames(logCPM)),
               row_split = factor(ord2, levels = lev2), 
               row_gap = unit(0, "cm"),
               width = unit(nr_degenes, "cm"), height = unit(10, "cm"),
               column_names_rot = 45)
               #rect_gp = gpar(col = "black", lwd = 1))


ht4 <- Heatmap(logchange_T, name = "Log2 FC",
               col = circlize::colorRamp2(c(-2, 0, 2), c("darkblue", "white", "green")),
               cluster_rows = F, cluster_columns = F, show_heatmap_legend = T,
               cell_fun = function(j, i, x, y, width, height, fill) {
                 grid.text(sprintf("%.1s", pvals_T[i, j]), x, y)
               }, 
               border = TRUE, height = unit(3, "cm"),
               row_names_max_width = max_text_width(c(0,0),gp = gpar(fontsize = 0)),
               column_names_rot = 45)
               #rect_gp = gpar(col = "black", lwd = 1))

ht_list = ht3 %v% ht4 
ht_list
svg("../analysis/heatmaps/heatmap_KFF_reduced.svg", width = nr_degenes, height = 10)
draw(ht_list, ht_gap = unit(0.5, "cm"))
dev.off()

```

# same for RXR:
```{r}

for (name in names(res_RXR)) {
  res_RXR[[name]]$table$p_value_FDR <- p.adjust(res_RXR[[name]]$table$PValue, method = "fdr")
  dataname <- paste("../analysis/", name, ".csv", sep = "")
  #write.csv(res_RXR[[name]]$table, dataname)
}

topDEgenes <- c(rownames(res_RXR$PNARXR$table[res_RXR$PNARXR$table$p_value_FDR<0.001 & 
                                           abs(res_RXR$PNARXR$table$logFC)>1,])[1:20],
                rownames(res_RXR$PNARXRscr$table[res_RXR$PNARXRscr$table$p_value_FDR<0.001 & 
                                           abs(res_RXR$PNARXRscr$table$logFC)>1,])[1:20],
                rownames(res_RXR$RXR$table[res_RXR$RXR$table$p_value_FDR<0.001 & 
                                           abs(res_RXR$RXR$table$logFC)>1,])[1:20], 
                "SL1344_1133", "SL1344_1134")
topDEgenes <- unique(topDEgenes[!is.na(topDEgenes)])


logCPM <- cpm(y, prior.count = 2, log = TRUE) 
logCPM <- logCPM[rownames(logCPM)%in%topDEgenes,]
logCPM <- t(scale(t(logCPM))) #centered around 0

#prefname <- ifelse(rownames(logCPM) %in% rownames(GO),GO[rownames(logCPM),]$Preferred_name, "" )
#rownames(logCPM) <- ifelse(prefname != "", prefname, rownames(logCPM))

logCPM <- logCPM[,RXR_nrs]

colnames(logCPM) <- factor(c("Untreated R1 ", "on-target PPNA R1 ","scrambled PPNA R1 ","peptide alone R1 ",
                             "Untreated R2 ", "on-target PPNA R2 ","scrambled PPNA R2 ","peptide alone R2 ",
                             "Untreated R3 ", "on-target PPNA R3 ","scrambled PPNA R3 ","peptide alone R3 "))

head(logCPM)
```

```{r}
library(ComplexHeatmap)
# get log2change, between those and noPNA:
logchange <- data.frame(PNARXR = res_RXR$PNARXR$table$logFC,PNARXRscr =res_RXR$PNARXRscr$table$logFC,
                              RXR =res_RXR$RXR$table$logFC, row.names = rownames(res_RXR$PNARXR$table))

pvals <- data.frame(PNARXR = (res_RXR$PNARXR$table$p_value_FDR<0.001 & abs(res_RXR$PNARXR$table$logFC)>1),
                        PNARXRscr=(res_RXR$PNARXRscr$table$p_value_FDR<0.001 & abs(res_RXR$PNARXRscr$table$logFC)>1),
                        RXR =(res_RXR$RXR$table$p_value_FDR<0.001 & abs(res_RXR$RXR$table$logFC)>1),
                        row.names = rownames(res_RXR$PNARXR$table))
#select only significant ones:
logchange <- logchange[rownames(logchange)%in%topDEgenes,]

colnames(logchange) <- factor(c("RXR-PNA ", "RXR-PNA-scr ", "RXR "))

pnames <- read.delim("../data/link_lt_gn.tab", header = F)
rownames(pnames) <- pnames$V2

prefname <- ifelse(rownames(logchange) %in% pnames$V2 ,pnames[rownames(logchange),]$V1, "" )
prefname <- ifelse(isUnique(prefname), prefname, "")
rownames(logchange) <- ifelse(prefname != "", prefname, rownames(logchange))



#select only significant ones:
pvals <- pvals[rownames(pvals)%in%topDEgenes,]
#rownames(pvals) <- x$Prefname
colnames(pvals) <- factor(c("on-target PPNA ", "scrambled PPNA ", "peptide alone "))
pvals <-sapply(pvals, function(x) ifelse(x, x <- "*",x<-"") )

#prefname <- ifelse(rownames(pvals) %in% rownames(GO),GO[rownames(pvals),]$Preferred_name, "" )
#rownames(pvals) <- ifelse(prefname != "", prefname, rownames(pvals))


colnames(logchange) <- factor(c("on-target PPNA ", "scrambled PPNA ", "peptide alone "))

pvals <- pvals[order(logchange$`on-target PPNA `, decreasing = T),]
logCPM <-logCPM[order(logchange$`on-target PPNA `, decreasing = T),]
logchange <- logchange[order(logchange$`on-target PPNA `, decreasing = T),]

logCPM_T <- t(logCPM)
logchange_T <- t(logchange)
pvals_T <- t(pvals)

rownames(logchange) <- gsub("(.*)", " \\1",rownames(logchange))
```


Now we save the heatmap as svg file:
```{r}

ord2 <- c(rep("on-target PPNA", 3), rep("peptide alone", 3), rep("scrambled PPNA", 3),
         rep("untreated", 3))
lev2 <- c("on-target PPNA", "scrambled PPNA", "peptide alone", "untreated")
logCPM_T <- logCPM_T[sort(rownames(logCPM_T)),]

nr_degenes <- dim(logCPM_T)[2]


ht3 <- Heatmap(logCPM_T, cluster_rows = F, name = "Log CPM",
               show_row_names = F,
               show_heatmap_legend = T, cluster_columns = F,
               row_title_side = "right", row_title_rot = 0,
               border = TRUE, 
               column_names_max_height=max_text_width(colnames(logCPM)),
               row_split = factor(ord2, levels = lev2), 
               row_gap = unit(0, "cm"),
               width = unit(nr_degenes, "cm"), height = unit(10, "cm"),
               column_names_rot = 45)
               #rect_gp = gpar(col = "black", lwd = 1))


ht4 <- Heatmap(logchange_T, name = "Log2 FC",
               col = circlize::colorRamp2(c(-2, 0, 2), c("darkblue", "white", "green")),
               cluster_rows = F, cluster_columns = F, show_heatmap_legend = T,
               cell_fun = function(j, i, x, y, width, height, fill) {
                 grid.text(sprintf("%.1s", pvals_T[i, j]), x, y)
               }, 
               border = TRUE, height = unit(3, "cm"),
               row_names_max_width = max_text_width(c(0,0),gp = gpar(fontsize = 0)),
               column_names_rot = 45)
               #rect_gp = gpar(col = "black", lwd = 1))

ht_list = ht3 %v% ht4 
ht_list
svg("../analysis/heatmaps/heatmap_RXR_reduced.svg", width = nr_degenes, height = 10)
draw(ht_list, ht_gap = unit(0.5, "cm"))
dev.off()

```
We can see that both of KFFs DE genes are also DE in both other comparisons





# same for TAT:
```{r}

for (name in names(res_TAT)) {
  res_TAT[[name]]$table$p_value_FDR <- p.adjust(res_TAT[[name]]$table$PValue, method = "fdr")
  dataname <- paste("../analysis/", name, ".csv", sep = "")
  #write.csv(res_TAT[[name]]$table, dataname)
}

topDEgenes <- c(rownames(res_TAT$PNATAT$table[res_TAT$PNATAT$table$p_value_FDR<0.001 & 
                                           abs(res_TAT$PNATAT$table$logFC)>1,])[1:20],
                rownames(res_TAT$PNATATscr$table[res_TAT$PNATATscr$table$p_value_FDR<0.001 & 
                                           abs(res_TAT$PNATATscr$table$logFC)>1,])[1:20],
                rownames(res_TAT$TAT$table[res_TAT$TAT$table$p_value_FDR<0.001 & 
                                           abs(res_TAT$TAT$table$logFC)>1,])[1:20], 
                "SL1344_1133", "SL1344_1134")
topDEgenes <- unique(topDEgenes[!is.na(topDEgenes)])



logCPM <- cpm(y, prior.count = 2, log = TRUE) 
logCPM <- logCPM[rownames(logCPM)%in%topDEgenes,]
logCPM <- t(scale(t(logCPM))) #centered around 0


#prefname <- ifelse(rownames(logCPM) %in% rownames(GO),GO[rownames(logCPM),]$Preferred_name, "" )
#rownames(logCPM) <- ifelse(prefname != "", prefname, rownames(logCPM))

logCPM <- logCPM[,TAT_nrs]

colnames(logCPM) <- factor(c("Untreated R1 ", "on-target PPNA R1 ","scrambled PPNA R1 ","peptide alone R1 ",
                             "Untreated R2 ", "on-target PPNA R2 ","scrambled PPNA R2 ","peptide alone R2 ",
                             "Untreated R3 ", "on-target PPNA R3 ","scrambled PPNA R3 ","peptide alone R3 "))

head(logCPM)
```

```{r}
library(ComplexHeatmap)
# get log2change, between those and noPNA:
logchange <- data.frame(PNATAT = res_TAT$PNATAT$table$logFC,PNATATscr =res_TAT$PNATATscr$table$logFC,
                              TAT =res_TAT$TAT$table$logFC, row.names = rownames(res_TAT$PNATAT$table))

pvals <- data.frame(PNATAT = (res_TAT$PNATAT$table$p_value_FDR<0.001 & abs(res_TAT$PNATAT$table$logFC)>1),
                        PNATATscr=(res_TAT$PNATATscr$table$p_value_FDR<0.001 & abs(res_TAT$PNATATscr$table$logFC)>1),
                        TAT =(res_TAT$TAT$table$p_value_FDR<0.001 & abs(res_TAT$TAT$table$logFC)>1),
                        row.names = rownames(res_TAT$PNATAT$table))
#select only significant ones:
logchange <- logchange[rownames(logchange)%in%topDEgenes,]

colnames(logchange) <- factor(c("TAT-PNA ", "TAT-PNA-scr ", "TAT "))


#pnames <- read.delim("../data/link_lt_gn.tab", header = F)
#rownames(pnames) <- pnames$V2
#prefname <- ifelse(rownames(logchange) %in% pnames$V2 ,pnames[rownames(logchange),]$V1, "" )
#prefname[134] <- ""
#rownames(logchange) <- ifelse(prefname != "", prefname, rownames(logchange))

prefname <- ifelse(rownames(logchange) %in% pnames$V2 ,pnames[rownames(logchange),]$V1, "" )
prefname <- ifelse(isUnique(prefname), prefname, "")
rownames(logchange) <- ifelse(prefname != "", prefname, rownames(logchange))




#select only significant ones:
pvals <- pvals[rownames(pvals)%in%topDEgenes,]
#rownames(pvals) <- x$Prefname
colnames(pvals) <- factor(c("on-target PPNA ", "scrambled PPNA ", "peptide alone "))
pvals <-sapply(pvals, function(x) ifelse(x, x <- "*",x<-"") )

#prefname <- ifelse(rownames(pvals) %in% rownames(GO),GO[rownames(pvals),]$Preferred_name, "" )
#rownames(pvals) <- ifelse(prefname != "", prefname, rownames(pvals))


colnames(logchange) <- factor(c("on-target PPNA ", "scrambled PPNA ", "peptide alone "))

pvals <- pvals[order(logchange$`on-target PPNA `, decreasing = T),]
logCPM <-logCPM[order(logchange$`on-target PPNA `, decreasing = T),]
logchange <- logchange[order(logchange$`on-target PPNA `, decreasing = T),]

logCPM_T <- t(logCPM)
logchange_T <- t(logchange)
pvals_T <- t(pvals)

rownames(logchange) <- gsub("(.*)", " \\1",rownames(logchange))
```


Now we save the heatmap as svg file:
```{r}

ord2 <- c(rep("on-target PPNA", 3), rep("peptide alone", 3), rep("scrambled PPNA", 3),
         rep("untreated", 3))
lev2 <- c("on-target PPNA", "scrambled PPNA", "peptide alone", "untreated")
logCPM_T <- logCPM_T[sort(rownames(logCPM_T)),]

nr_degenes <- dim(logCPM_T)[2]

ht3 <- Heatmap(logCPM_T, cluster_rows = F, name = "Log CPM",
               show_row_names = F,
               show_heatmap_legend = T, cluster_columns = F,
               row_title_side = "right", row_title_rot = 0,
               border = TRUE, 
               column_names_max_height=max_text_width(colnames(logCPM)),
               row_split = factor(ord2, levels = lev2), 
               row_gap = unit(0, "cm"),
               width = unit(nr_degenes, "cm"), height = unit(10, "cm"),
               column_names_rot = 45)
               #rect_gp = gpar(col = "black", lwd = 1))


ht4 <- Heatmap(logchange_T, name = "Log2 FC",
               col = circlize::colorRamp2(c(-2, 0, 2), c("darkblue", "white", "green")),
               cluster_rows = F, cluster_columns = F, show_heatmap_legend = T,
               cell_fun = function(j, i, x, y, width, height, fill) {
                 grid.text(sprintf("%.1s", pvals_T[i, j]), x, y)
               }, 
               border = TRUE, height = unit(3, "cm"),
               row_names_max_width = max_text_width(c(0,0),gp = gpar(fontsize = 0)),
               column_names_rot = 45)
               #rect_gp = gpar(col = "black", lwd = 1))

ht_list = ht3 %v% ht4 
ht_list
svg("../analysis/heatmaps/heatmap_TAT_reduced.svg", width = nr_degenes, height = 10)
draw(ht_list, ht_gap = unit(0.5, "cm"))
dev.off()

```

#Venn diagrams:
```{r}
library(VennDiagram)
library(eulerr)
library(tidyverse)
library(grid)


list_DEgenes <- list(PNA_TAT = rownames(res_TAT$PNATAT$table[res_TAT$PNATAT$table$p_value_FDR<0.01 & 
                                           abs(res_TAT$PNATAT$table$logFC)>1,]),
                PNA_TAT_scr = rownames(res_TAT$PNATATscr$table[res_TAT$PNATATscr$table$p_value_FDR<0.01 & 
                                           abs(res_TAT$PNATATscr$table$logFC)>1,]),
                TAT = rownames(res_TAT$TAT$table[res_TAT$TAT$table$p_value_FDR<0.01 & 
                                           abs(res_TAT$TAT$table$logFC)>1,]),
                PNA_RXR = rownames(res_RXR$PNARXR$table[res_RXR$PNARXR$table$p_value_FDR<0.01 & 
                                           abs(res_RXR$PNARXR$table$logFC)>1,]),
                PNA_RXR_scr = rownames(res_RXR$PNARXRscr$table[res_RXR$PNARXRscr$table$p_value_FDR<0.01 & 
                                           abs(res_RXR$PNARXRscr$table$logFC)>1,]),
                RXR = rownames(res_RXR$RXR$table[res_RXR$RXR$table$p_value_FDR<0.01 & 
                                           abs(res_RXR$RXR$table$logFC)>1,]),
                PNA_KFF = rownames(res_KFF$PNAKFF$table[res_KFF$PNAKFF$table$p_value_FDR<0.01 & 
                                           abs(res_KFF$PNAKFF$table$logFC)>1,]),
                PNA_KFF_scr = rownames(res_KFF$PNAKFFscr$table[res_KFF$PNAKFFscr$table$p_value_FDR<0.01 & 
                                           abs(res_KFF$PNAKFFscr$table$logFC)>1,]),
                KFF = rownames(res_KFF$KFF$table[res_KFF$KFF$table$p_value_FDR<0.01 & 
                                           abs(res_KFF$KFF$table$logFC)>1,]))

svg("../analysis/VennDiagrams/Venn_TMM_TAT",width = 10, height = 10)
plot(euler(list_DEgenes[1:3]) , quantities = T)
dev.off()

svg("../analysis/VennDiagrams/Venn_TMM_RXR",width = 10, height = 10)
plot(euler(list_DEgenes[4:6]) , quantities = T)
dev.off()

svg("../analysis/VennDiagrams/Venn_TMM_KFF",width = 10, height = 10)
plot(euler(list_DEgenes[7:9]) , quantities = T)
dev.off()

svg("../analysis/VennDiagrams/Venn_TMM_peptides",width = 10, height = 10)
plot(euler(list_DEgenes[c(3,6,9)]) , quantities = T)
dev.off()

svg("../analysis/VennDiagrams/Venn_TMM_PPNAs",width = 10, height = 10)
plot(euler(list_DEgenes[c(1,4,7)]) , quantities = T)
dev.off()
```


# Gene Ontology


Now we import preferred gene names etc from the egnog annotation:
```{r}
# read in eggnog file
file <- "../data/EggNogg/query_seqs.fa.emapper.annotations"
GO <- read.delim(file ,sep = "\t",  header = T,row.names = 1, comment.char = "#")
rownames(GO) <- gsub("gene-", "", rownames(GO))
head(GO)
```


GO-fry analysis:
```{r}
library(GO.db)
#get GO terms:
gene_GO_links <- data.frame(gene = rownames(GO), GOid = GO$GOs)
gene_GO_links <- gene_GO_links[rownames(GO) %in% rownames(res_KFF$PNAKFF$table),]

gene_GO_links$GOid <- as.character(gene_GO_links$GOid)  
all_pw_ids <- unique(unlist(strsplit(gene_GO_links$GOid,",")))
list_GO_ids <- strsplit(gene_GO_links$GOid,",")

idx <- lapply(all_pw_ids, function(x){
  tf <- unlist(lapply(list_GO_ids, function(y) x %in% y))
  x <- unique(gene_GO_links$gene[tf]) # choose all genes, except duplucates
})
names(idx) <- all_pw_ids

# add phopq pw to gos
ppq_raw <- read.delim("../data/PHOPQ.tsv", header = F)
ppq <- as.character(ppq_raw$V1)
phopq <- pnames[pnames$V1 %in% ppq,]$V2

idx$PhoPQ <- phopq[phopq %in% rownames(y$counts)]


#do fry: 
fry_PNAKFF <- fry(y, idx, design, contrast=con[,1])
fry_KFF <- fry(y, idx, design, contrast=con[,3])
fry_PNAKFFscr <- fry(y, idx, design, contrast=con[,2])

fry_PNARXR <- fry(y, idx, design, contrast=con[,4])
fry_RXR <- fry(y, idx, design, contrast=con[,6])
fry_PNARXRscr <- fry(y, idx, design, contrast=con[,5])

fry_PNATAT <- fry(y, idx, design, contrast=con[,7])
fry_TAT <- fry(y, idx, design, contrast=con[,9])
fry_PNATATscr <- fry(y, idx, design, contrast=con[,8])

list_fry <- list(fry_PNAKFF=fry_PNAKFF,fry_PNAKFFscr=fry_PNAKFFscr,fry_KFF=fry_KFF,
                 fry_PNARXR=fry_PNARXR,fry_PNARXRscr=fry_PNARXRscr,fry_RXR=fry_RXR,
                 fry_PNATAT=fry_PNATAT,fry_PNATATscr=fry_PNATATscr,fry_TAT=fry_TAT)


index <- rownames(fit) %in% phopq
barcodeplot(res_KFF$KFF$table$logFC, index, 
            labels = c("Control", "PPNA"), main = "PhoPQ")

```

```{r}
link_GO_terms_fry <- select(GO.db, keys = rownames(fry_PNAKFF), columns = "TERM")
rownames(link_GO_terms_fry) <- link_GO_terms_fry$GOID

fry_PNAKFF$TERM <-  link_GO_terms_fry[rownames(fry_PNAKFF),]$TERM
fry_PNAKFFscr$TERM <-  link_GO_terms_fry[rownames(fry_PNAKFFscr),]$TERM
fry_KFF$TERM <-  link_GO_terms_fry[rownames(fry_KFF),]$TERM

for (fryres in names(list_fry)) {
  list_fry[[fryres]][["TERM"]] <- link_GO_terms_fry[rownames(list_fry[[fryres]]),][["TERM"]]
  list_fry[[fryres]]["PhoPQ",][["TERM"]] <- "PhoPQ"
}

frysig <- lapply(list_fry, function(x) x[x[["FDR"]]<0.05 & x[["NGenes"]]>2,])

siggos <- c()

for (i in names(frysig)) {
  print(i)
  print(dim(frysig[[i]]))
  print(frysig[[i]][,c(1,2,4,7)])
  siggos <- c(siggos, rownames(frysig[[i]][1:10,]))  # we could change that
}

siggos <- unique(siggos[!grepl("NA", siggos)])
```

Create a heatmap-df  for GO-terms:
```{r}
idx_char <- lapply(idx, as.character)
# I create a dataframe with mean logFC values for each significant GO-term:
hm_fry_logfc <- t(as.data.frame(lapply(idx_char[siggos], function(x){
  PNAKFF <- median(res_KFF$PNAKFF$table[x,]$logFC)
  PNAKFFscr <- median(res_KFF$PNAKFFscr$table[x,]$logFC)
  KFF <- median(res_KFF$KFF$table[x,]$logFC)
  
  PNARXR <- median(res_RXR$PNARXR$table[x,]$logFC)
  PNARXRscr <- median(res_RXR$PNARXRscr$table[x,]$logFC)
  RXR <- median(res_RXR$RXR$table[x,]$logFC)
  
  PNATAT <- median(res_TAT$PNATAT$table[x,]$logFC)
  PNATATscr <- median(res_TAT$PNATATscr$table[x,]$logFC)
  TAT <- median(res_TAT$TAT$table[x,]$logFC)
  c(PNAKFF, PNAKFFscr, KFF, PNARXR, PNARXRscr, RXR, PNATAT, PNATATscr, TAT)
})))

colnames(hm_fry_logfc) <- c("KFF-acpP ", "KFF-acpP-scrambled ", "KFF-only ","RXR-acpP ", "RXR-acpP-scrambled ", "RXR-only ",
                            "TAT-acpP ", "TAT-acpP-scrambled ", "TAT-only ")
hm_fry_logfc_TAT <- as.data.frame(hm_fry_logfc)
rownames(hm_fry_logfc) <- gsub("\\.", "\\:", rownames(hm_fry_logfc))
```

plot heatmap:
```{r}
hm_fry_logfc <- hm_fry_logfc[order(hm_fry_logfc[,1], decreasing = T),]

pvals <- data.frame(sapply(names(list_fry), function(x) list_fry[[x]][rownames(hm_fry_logfc),"FDR"]), 
                    row.names = rownames(hm_fry_logfc))

#select only significant ones:
pvals <-sapply(pvals, function(x) ifelse(x<0.05, x <- "*",x<-"") )


goterms <- list_fry$fry_PNAKFF[rownames(hm_fry_logfc),] [["TERM"]]
rownames(hm_fry_logfc) <- ifelse(!is.na(goterms),goterms, rownames(hm_fry_logfc) )

seper <- c(1) 
for (i in 2:dim(hm_fry_logfc)[1]) {
  ifelse(sum(hm_fry_logfc[i,]) == sum(hm_fry_logfc[i-1,]),
         seper <- c(seper,seper[i-1]), seper <- c(seper,i))
}


ord <- c(rep("KFF", 3), rep("RXR", 3), rep("TAT", 3))
lev <- c("KFF", "RXR", "TAT")
```

make hm:
```{r}
col_fun = colorRamp2(c(-1, 0, 1), c("darkblue", "beige", "red"))

w <- length(rownames(hm_fry_logfc) )# width of plot (nr pws)
h <- 9   # height of plot 

ht_vert <- Heatmap(hm_fry_logfc, cluster_rows = F, cluster_columns = F,
               name = "GO-analysis", col = col_fun,
               show_heatmap_legend = F, 
               row_title_side = "right", row_title_rot = 0,
               border = TRUE, 
               cell_fun = function(j, i, x, y, width, height, fill) {
                 grid.text(sprintf("%.1s", pvals[i, j]), x, y)
               }, 
               column_names_gp = gpar(fontsize = 10),
               row_names_gp = gpar(fontsize = 10),
               column_split = factor(ord, levels = lev), 
               row_split = factor(seper), 
               row_title = NULL,
               row_gap = unit(0.1, "cm"),
               width = unit(5, "cm"), height = unit(20, "cm"),
               column_names_rot = 45,
               rect_gp = gpar(col = "black", lwd = 0.1))

ht_hor <- Heatmap(t(hm_fry_logfc), cluster_rows = F, cluster_columns = F,
        name = "GO-analysis", col = col_fun,
        show_heatmap_legend = F, 
        row_title_side = "left", row_title_rot = 0,
        column_names_side = "top",
        border = TRUE, 
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sprintf("%.1s", t(pvals)[i, j]), x, y)
        }, 
        row_names_gp = gpar(fontsize = 10),
        column_names_gp = gpar(fontsize = 10),
        row_split = factor(ord, levels = lev), 
        column_split = factor(seper), 
        column_title = NULL,
        column_gap = unit(0.1, "cm"),
        width = unit(w*0.8, "cm"), height = unit(h*0.8, "cm"),
        column_names_rot = 40,
        rect_gp = gpar(col = "black", lwd = 0.1))

lgd = Legend(col_fun = col_fun, title = "median logFC", direction = "horizontal",
             at = c(-1, 0, 1), legend_width = unit(4, "cm"))

svg("../analysis/gene_set_analysis/hm_GO_collapsed.svg", width = unit(0.45*w, "cm"))
draw(ht_hor)
draw(lgd, x = unit(0.45*w, "cm"), y = unit(0.8, "cm"), just = c("left", "bottom"))
dev.off()
```




KEGG-analysis:
```{r}
library(KEGGREST)
# get link and list to get kegg info:
link_kegg <- keggLink("pathway", "sey")
list_kegg <- keggList("pathway", "sey")

kegg_pw_ids <- names(list_kegg)

#rename genes, remove ones which arent in our data:
names(link_kegg) <- gsub("sey:(.*)", "\\1", names(link_kegg)) #rename genes as locus tags
link_kegg <- link_kegg[names(link_kegg) %in% rownames(res_KFF$PNAKFF$table)] #remove genes not in data


idx_kegg <- sapply(kegg_pw_ids, function(x){
  x <- unique(names(link_kegg[link_kegg == x])) # choose all genes, except duplucates
})
idx_kegg$PhoPQ <- phopq[phopq %in% rownames(y$counts)] # add PhoPQ genes


#do fry: 
kegg_fry_PNAKFF <- fry(y, idx_kegg, design, contrast=con[,1])
kegg_fry_KFF <- fry(y, idx_kegg, design, contrast=con[,3])
kegg_fry_PNAKFFscr <- fry(y, idx_kegg, design, contrast=con[,2])

kegg_fry_PNARXR <- fry(y, idx_kegg, design, contrast=con[,4])
kegg_fry_RXR <- fry(y, idx_kegg, design, contrast=con[,6])
kegg_fry_PNARXRscr <- fry(y, idx_kegg, design, contrast=con[,5])

kegg_fry_PNATAT <- fry(y, idx_kegg, design, contrast=con[,7])
kegg_fry_TAT <- fry(y, idx_kegg, design, contrast=con[,9])
kegg_fry_PNATATscr <- fry(y, idx_kegg, design, contrast=con[,8])

list_kegg_fry <- list(kegg_fry_PNAKFF=kegg_fry_PNAKFF,kegg_fry_PNAKFFscr=kegg_fry_PNAKFFscr,kegg_fry_KFF=kegg_fry_KFF,
                 kegg_fry_PNARXR=kegg_fry_PNARXR,kegg_fry_PNARXRscr=kegg_fry_PNARXRscr,kegg_fry_RXR=kegg_fry_RXR,
                 kegg_fry_PNATAT=kegg_fry_PNATAT,kegg_fry_PNATATscr=kegg_fry_PNATATscr,kegg_fry_TAT=kegg_fry_TAT)

```

add terms:
```{r}
for (fryres in names(list_kegg_fry)) {
  list_kegg_fry[[fryres]][["TERM"]] <- list_kegg[rownames(list_kegg_fry[[fryres]])]
  list_kegg_fry[[fryres]][["TERM"]] <- gsub("(.*) - Salmonella enterica subsp. enterica serovar Typhimurium SL1344",
                                            "\\1", list_kegg_fry[[fryres]][["TERM"]])
  list_kegg_fry[[fryres]]["PhoPQ",][["TERM"]] <- "PhoPQ"
}


kegg_frysig <- lapply(list_kegg_fry, function(x) x[x[["FDR"]]<0.05 & x[["NGenes"]]>10,])
kegg_siggos <- c()


for (i in names(kegg_frysig)) {
  print(i)
  print(dim(kegg_frysig[[i]]))
  print(kegg_frysig[[i]][,c(1,2,4,7)])
  kegg_siggos <- c(kegg_siggos, rownames(kegg_frysig[[i]][,]))  # can be modified
}

kegg_siggos <- unique(kegg_siggos[!grepl("NA", kegg_siggos)])


```

Create a heatmap-df  for KEGG:
```{r}
idx_kegg_char <- lapply(idx_kegg, as.character)
# I create a dataframe with mean logFC values for each significant GO-term:
hm_kegg_fry_logfc <- t(as.data.frame(lapply(idx_kegg_char[kegg_siggos], function(x){
  PNAKFF <- median(res_KFF$PNAKFF$table[x,]$logFC)
  PNAKFFscr <- median(res_KFF$PNAKFFscr$table[x,]$logFC)
  KFF <- median(res_KFF$KFF$table[x,]$logFC)
  
  PNARXR <- median(res_RXR$PNARXR$table[x,]$logFC)
  PNARXRscr <- median(res_RXR$PNARXRscr$table[x,]$logFC)
  RXR <- median(res_RXR$RXR$table[x,]$logFC)
  
  PNATAT <- median(res_TAT$PNATAT$table[x,]$logFC)
  PNATATscr <- median(res_TAT$PNATATscr$table[x,]$logFC)
  TAT <- median(res_TAT$TAT$table[x,]$logFC)
  c(PNAKFF, PNAKFFscr, KFF, PNARXR, PNARXRscr, RXR, PNATAT, PNATATscr, TAT)
})))

colnames(hm_kegg_fry_logfc) <- c("KFF-acpP ", "KFF-acpP-scrambled ", "KFF-only ","RXR-acpP ", "RXR-acpP-scrambled ", "RXR-only ",
                            "TAT-acpP ", "TAT-acpP-scrambled ", "TAT-only ")
hm_kegg_fry_logfc <- as.data.frame(hm_kegg_fry_logfc)
rownames(hm_kegg_fry_logfc) <- gsub("\\.", "\\:", rownames(hm_kegg_fry_logfc))
```

make heatmap:
```{r}
hm_kegg_fry_logfc <- hm_kegg_fry_logfc[order(hm_kegg_fry_logfc[,1], decreasing = T),]

pvals <- data.frame(sapply(names(list_kegg_fry), function(x) list_kegg_fry[[x]][rownames(hm_kegg_fry_logfc),"FDR"]), 
                    row.names = rownames(hm_kegg_fry_logfc))

#select only significant ones:
pvals <-sapply(pvals, function(x) ifelse(x<0.05, x <- "*",x<-"") )


keggpws <- list_kegg_fry$kegg_fry_PNAKFF[rownames(hm_kegg_fry_logfc),] [["TERM"]]
rownames(hm_kegg_fry_logfc) <- ifelse(!is.na(keggpws),keggpws, rownames(hm_kegg_fry_logfc) )


ord <- c(rep("KFF", 3), rep("RXR", 3), rep("TAT", 3))
lev <- c("KFF", "RXR", "TAT")
```

plot hm:
```{r}
library(circlize)
col_fun = colorRamp2(c(-1, 0, 1), c("darkblue", "beige", "red"))

w <- length(hm_kegg_fry_logfc$`KFF-acpP `) # width of plot (nr pws)
h <- 9   # height of plot 

ht_vert <- Heatmap(hm_kegg_fry_logfc, cluster_rows = F, cluster_columns = F,
               name = "GO-analysis", col = col_fun,
               show_heatmap_legend = F, 
               row_title_side = "right", row_title_rot = 0,
               border = TRUE, 
               cell_fun = function(j, i, x, y, width, height, fill) {
                 grid.text(sprintf("%.1s", pvals[i, j]), x, y)
               }, 
               column_names_gp = gpar(fontsize = 10),
               row_names_gp = gpar(fontsize = 10),
               column_split = factor(ord, levels = lev), 
               row_title = NULL,
               row_gap = unit(0.1, "cm"),
               width = unit(5, "cm"), height = unit(20, "cm"),
               column_names_rot = 45,
               rect_gp = gpar(col = "black", lwd = 0.1))

ht_hor <- Heatmap(t(hm_kegg_fry_logfc), cluster_rows = F, cluster_columns = F,
        name = "GO-analysis", col = col_fun,
        show_heatmap_legend = F, 
        row_title_side = "left", row_title_rot = 0,
        column_names_side = "top",
        border = TRUE, 
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sprintf("%.1s", t(pvals)[i, j]), x, y)
        }, 
        row_names_gp = gpar(fontsize = 10),
        column_names_gp = gpar(fontsize = 10),
        row_split = factor(ord, levels = lev), 
        column_title = NULL,
        column_gap = unit(0.1, "cm"),
        width = unit(w*0.8, "cm"), height = unit(h*0.8, "cm"),
        column_names_rot = 40,
        rect_gp = gpar(col = "black", lwd = 0.1))

lgd = Legend(col_fun = col_fun, title = "median logFC", direction = "horizontal",
             at = c(-1, 0, 1), legend_width = unit(4, "cm"))

svg("../analysis/gene_set_analysis/hm_KEGG_complete.svg", width = unit(w*0.45, "cm"))
draw(ht_hor)
draw(lgd, x = unit(w*0.45, "cm"), y = unit(0.8, "cm"), just = c("left", "bottom"))
dev.off()
```


# SPI2 analysis:
load data:
```{r}
df_spi2 <- read.delim("../data/PhoPQ_analysis_salcom/salcom_spi2genes_colgan.txt", header = T, sep="\t")
df_phopq <- read.delim("../data/PhoPQ_analysis_salcom/salcom_query.phopq.txt", header = T, sep="\t")

df_salcom_spi2 <- data.frame(Wildtype = df_spi2$WT.MEP, SPI2 = df_spi2$WT.InSPI2, 
                        PhoPQ_ko = df_spi2$X.Delta.phoP.Q.InSPI2, row.names = df_spi2$SL1344.Locus.ID)

df_salcom_phopq <- data.frame(Wildtype = df_phopq$WT.MEP, SPI2 = df_phopq$WT.InSPI2, 
                        PhoPQ_ko = df_phopq$X.Delta.phoP.Q.InSPI2, row.names = df_phopq$SL1344.Locus.ID)

tpm_salcom_spi2 <- data.frame(sapply(df_salcom_spi2, function(x) as.integer(gsub(",","", x))), 
                              row.names = rownames(df_salcom_spi2))
tpm_salcom_phopq <- data.frame(sapply(df_salcom_phopq, function(x) as.integer(gsub(",","", x))), 
                              row.names = rownames(df_salcom_phopq))

tpm_salcom_spi2 <- tpm_salcom_spi2[order(tpm_salcom_spi2$SPI2),]
tpm_salcom_phopq <- tpm_salcom_phopq[order(tpm_salcom_phopq$SPI2),]
tpm_salcom_spi2 <- tpm_salcom_spi2[!(rownames(tpm_salcom_spi2)=="SL1344_1325"),]

tpm_salcom_spi2$condition <- "SPI2"
tpm_salcom_phopq$condition <- "PhoPQ"

rownames(tpm_salcom_phopq) <- gsub(".*MgrR", "MgrR", rownames(tpm_salcom_phopq))
tpm_salcom <- data.frame(rbind(tpm_salcom_phopq, tpm_salcom_spi2), row.names = c(rownames(tpm_salcom_phopq),
                                                                                 rownames(tpm_salcom_spi2)))

str(tpm_salcom)
```
Get all genes which are in our dataset:
```{r}
all_spi2_genes <- rownames(tpm_salcom)[rownames(tpm_salcom) %in% rownames(y)]
str(all_spi2_genes)
tpm_salcom <- tpm_salcom[all_spi2_genes,]

prefname <- ifelse(all_spi2_genes %in% pnames$V2 ,pnames[all_spi2_genes,]$V1, "" )
prefname <- ifelse(isUnique(prefname), prefname, "")
prefname_spi2 <- ifelse(prefname != "", prefname, all_spi2_genes)

rownames(tpm_salcom) <- prefname_spi2
```

Now we have to make heatmaps with log foldchanges for all samples:
```{r}
# make list of all results, get logchange table:
reslist <- list(KFF = res_KFF,RXR = res_RXR, TAT = res_TAT)
logchanges_spi2 <- data.frame(lapply(reslist, function(l) {
  logfc <- data.frame( sapply(names(l), function(r) {
    l[[r]]$table[all_spi2_genes,1]
  }), row.names = all_spi2_genes )
})) 

colnames(logchanges_spi2) <- c("KFF-acpP ", "KFF-acpP-scrambled ", "KFF-only ","RXR-acpP ",
                          "RXR-acpP-scrambled ", "RXR-only ",
                          "TAT-acpP ", "TAT-acpP-scrambled ", "TAT-only")


# get mean cpm values of all conditions:
countscpm <- cpm(y)[all_spi2_genes,]
#levels(test) <- levels(test)[c(2,3,1,5,6,4,8,9,7,10)]
spi2_cpm <- sapply(levels(test), function(t) {
  rowMeans(countscpm[,t == test])
})
spi2_cpm <- spi2_cpm[,c(2,3,1,5,6,4,8,9,7,10)]


pvalues_spi2 <- data.frame(lapply(reslist, function(l) {
  logfc <- data.frame( sapply(names(l), function(r) {
    l[[r]]$table[all_spi2_genes,]$p_value_FDR<0.001 & abs(l[[r]]$table[all_spi2_genes,]$logFC)>1
  }), row.names = all_spi2_genes )
})) 

pvalues_spi2 <-sapply(pvalues_spi2, function(x) ifelse(x , x <- "*",x<-"")) 



```




Plot Heatmap:
```{r}
h <- dim(logchanges_spi2)[1] # width of plot (nr pws)

col_fun = colorRamp2(c(-2, 0, 2), c("darkblue", "beige", "red"))
ht_vert <- Heatmap(logchanges_spi2, cluster_rows = F, cluster_columns = F,
               name = "SPI2 analysis", col = col_fun,
               show_heatmap_legend = F, 
               row_title_side = "left", row_title_rot = 0,
               border = TRUE, 
               cell_fun = function(j, i, x, y, width, height, fill) {
                 grid.text(sprintf("%.1s", pvalues_spi2[i, j]), x, y)
               }, 
               column_names_gp = gpar(fontsize = 10),
               row_names_gp = gpar(fontsize = 10),
               column_split = factor(ord, levels = lev), 
               row_split = factor(tpm_salcom$condition),
               row_gap = unit(0.2, "cm"),
               width = unit(9*0.7, "cm"), height = unit(h/2, "cm"),
               column_names_rot = 45,
               rect_gp = gpar(col = "black", lwd = 0.1))


col_col_fun = colorRamp2(c(0, 2,3), c("lightgrey", "yellow", "red"))
ht_colgan <- Heatmap(log10(tpm_salcom[,1:3]), cluster_rows = F, cluster_columns = F,
               name = "Colgan et al., 2016", col = col_col_fun,
               show_heatmap_legend = F, 
               row_title_side = "right", row_title_rot = 0,
               border = TRUE, 
               row_split = factor(tpm_salcom$condition),
               column_names_gp = gpar(fontsize = 10),
               row_names_gp = gpar(fontsize = 10),
               row_gap = unit(0.2, "cm"),
               width = unit(3*0.7, "cm"), height = unit(h/2, "cm"),
               column_names_rot = 45,
               rect_gp = gpar(col = "black", lwd = 0.1))

col_fun_cpm = colorRamp2(c(0, 1.7, 2), c("lightgrey", "yellow", "red"))
order = factor(c(rep("KFF", 3), rep("RXR", 3), rep("TAT", 3), "Control"))
ht_cpm <- Heatmap(log10(spi2_cpm), cluster_rows = F, cluster_columns = F,
        name = "SPI2 analysis", col = col_fun_cpm,
        show_heatmap_legend = F, 
        row_title_side = "right", row_title_rot = 0,
        border = TRUE, 
        column_names_gp = gpar(fontsize = 10),
        row_names_gp = gpar(fontsize = 10),
        column_split = factor(order), 
        row_title = NULL,
        row_gap = unit(0.1, "cm"),
        width = unit(9*0.7, "cm"), height = unit(h/2, "cm"),
        column_names_rot = 45,
        rect_gp = gpar(col = "black", lwd = 0.1))


ht_list =  ht_vert + ht_colgan 
ht_list

lgd = Legend(col_fun = col_fun, title = "Mean log2FC",
             at = c(-2, 0, 2), legend_width = unit(4, "cm"),
             labels = c("-2", " 0", " 2"), legend_height = unit(3, "cm"),
             title_position = "leftcenter-rot")
lgd_col = Legend(col_fun = col_col_fun, title = "Transcripts per million",
             at = c(0, 1, 2, 3), labels = c(0,10,100,1000), legend_width = unit(3, "cm"),
             legend_height = unit(3, "cm"),
             title_position = "leftcenter-rot")


svg("../analysis/PhoPQ_Salcom/hm_salcom_phopq_spi2.svg", width = 10, height = 15)
draw(ht_list, ht_gap = unit(0.5, "cm"))
draw(lgd, x = unit(4, "cm"), y = unit(18, "cm"), just = c("left", "bottom"))
draw(lgd_col, x = unit(20, "cm"), y = unit(18, "cm"), just = c("left", "bottom"))
dev.off()
```

```{r}
for (resname in names(all_res)){
  for (name in names(all_res[[resname]])){
    rtnew <- all_res[[resname]][[name]]$table
    rtold <- old_reslist[[resname]][[name]]$table
    rtold <- rtold[rownames(rtold)%in%rownames(rtnew),]
    rtnew <- rtnew[rownames(rtnew)%in%rownames(rtold),]
    rtold <- rtold[rownames(rtnew),]
    
    prefname <- ifelse(rownames(rtnew) %in% pnames$V2 ,pnames[rownames(rtnew),]$V1, "" )
    prefname <- ifelse(isUnique(prefname), prefname, "")
    cnames <- ifelse(prefname != "", prefname, rownames(rtnew))
    
    

    
    d <- data.frame(new_logFC = rtnew$logFC,old_logfc = rtold$logFC, change = abs(rtold$logFC-rtnew$logFC),
                    new_pval = rtnew$FDR, old_pval = p.adjust(rtold$PValue, method = "fdr"),
                    row.names = cnames)
    
    old_DE <- rownames(d[(d$old_logfc>1|d$old_logfc<(-1)) & d$old_pval< 0.001,])
    new_DE <- rownames(d[(d$new_logFC>1|d$new_logFC<(-1)) & d$new_pval< 0.001,])
    both_DE <- old_DE[old_DE %in% new_DE]
    
    on <- unique(c(new_DE, old_DE))
    on <- on[!(on %in% both_DE)]
    topon <- d[on,]
    topon <- topon[order(topon$change, decreasing = TRUE),]
    topon <- rownames(head(topon, n=20))
    
    cols <- rep("not DE", length(cnames))
    cols[(d$old_logfc>1|d$old_logfc<(-1)) & d$old_pval< 0.001] <- "old DE"
    cols[(d$new_logFC>1|d$new_logFC<(-1)) & d$new_pval< 0.001] <- "new DE"
    cols[((d$new_logFC>1|d$new_logFC<(-1)) & d$new_pval< 0.001)|
           ((d$old_logfc>1|d$old_logfc<(-1)) & d$old_pval< 0.001)] <- "both DE"
    d$DE <- factor(cols)
    
    g <- ggplot(d, aes(x = new_logFC, y = old_logfc),color = "black",cex = 2) + 
      geom_point() + stat_smooth(method=lm) + ggtitle(name) + xlim(-3,7)+ ylim(-3,7)+
      theme_bw()
  
    colors <- c("old DE" = "blue", "new DE" = "red", "both DE" = "green")
    g <- g + geom_point(
          data = d[old_DE,],
          aes(x = new_logFC, y = old_logfc),
          color = colors["old DE"], 
          cex = 2) +
      geom_point(
          data = d[new_DE,],
          aes(x = new_logFC, y = old_logfc),
          color = colors["new DE"], 
          cex = 2) +
      geom_point(
          data = d[both_DE,],
          aes(x = new_logFC, y = old_logfc),
          color = colors["both DE"], 
          cex = 2) 
    
    g <- g +geom_label_repel(  data = d[topon,] ,
     aes(x = new_logFC, y = old_logfc,label = rownames(d[topon,])),
      hjust = 0.1,
      vjust = 2,
      size = 4, segment.alpha = 0.5,min.segment.length=0, segment.color = "black")
    
    print(g)
    svg(paste("../analysis/comparison_experiments/", name, ".svg", sep = ""))
    print(g)
    dev.off()
  }
}
```


Packages used:
```{r}
sessionInfo()
```

